<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Object QC Functions • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Object QC Functions">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Object_QC_Functions_files/kePrint-0.0.1/kePrint.js"></script><link href="Object_QC_Functions_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Object QC Functions</h1>
            
            <h4 data-toc-skip class="date">Compiled: December 19,
2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/articles/Object_QC_Functions.Rmd" class="external-link"><code>vignettes/articles/Object_QC_Functions.Rmd</code></a></small>
      <div class="hidden name"><code>Object_QC_Functions.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="quality-control-of-scrna-seq-objects">Quality control of scRNA-seq objects<a class="anchor" aria-label="anchor" href="#quality-control-of-scrna-seq-objects"></a>
</h2>
<p>scCustomize has several helper functions to simplify/streamline what
is nearly always the first and most critical choices when starting an
analysis: performing quality control and filtering.</p>
<p>Let’s load packages and raw data object for this tutorial.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">pbmc3k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/pbmc3k.SeuratData/man/pbmc3k.html" class="external-link">pbmc3k</a></span></span></code></pre></div>
<p>We’ll add some random meta data variables to pbmc data form use in
this vignette</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span>, <span class="st">"sample4"</span>, <span class="st">"sample5"</span>, <span class="st">"sample6"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,</span>
<span>    replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Batch1"</span>, <span class="st">"Batch2"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="cross-compatibility-of-qc-functions">Cross-compatibility of QC Functions<a class="anchor" aria-label="anchor" href="#cross-compatibility-of-qc-functions"></a>
</h3>
<div class="section level4">
<h4 id="support-for-seurat-and-liger-objects">Support for Seurat and LIGER Objects<a class="anchor" aria-label="anchor" href="#support-for-seurat-and-liger-objects"></a>
</h4>
<p>All of scCustomize’s functions to add quality control metrics are
100% cross compatible across Seurat and LIGER objects using the same
function calls. For more details on QC specific plotting functions see
<a href="https://samuel-marsh.github.io/scCustomize/articles/QC_Plots.html">QC
Plotting &amp; Analysis Vignette</a>.</p>
</div>
<div class="section level4">
<h4 id="support-for-gene-symbols-and-ensembl-ids">Support for gene symbols and Ensembl IDs<a class="anchor" aria-label="anchor" href="#support-for-gene-symbols-and-ensembl-ids"></a>
</h4>
<p>Additionally, all of the QC functions support objects that use either
gene symbols or Ensembl IDs. Ensembl IDs for default species (see below)
are from Ensembl version 112 (updated in scCustomize on 4/29/2024).</p>
<p>If your object using ensembl IDs as features names then simply add
<code>ensembl_ids</code> parameter that is present in all QC
functions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using gene name patterns</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, species <span class="op">=</span> <span class="st">"Human"</span>, ensembl_ids <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="default-species-support">Default Species Support<a class="anchor" aria-label="anchor" href="#default-species-support"></a>
</h3>
<p>Many of the QC functions commonly performed depend on genes within a
particular family that have similar naming patterns (e.g., Mitochondrial
genes) or are species specific (see msigdb dependent parts of
<code><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics()</a></code>).</p>
<p>To simplify the process of needing to remember species-specific
patterns (or find Ensembl ID gene lists)</p>
<p>If you are using mouse, human, marmoset, zebrafish, rat, drosophila,
rhesus macaque, or chicken data all you need to do is specify the
<code>species</code> parameter in the functions described below using
one of the following accepted names.</p>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Mouse_Options
</th>
<th style="text-align:left;">
Human_Options
</th>
<th style="text-align:left;">
Marmoset_Options
</th>
<th style="text-align:left;">
Zebrafish_Options
</th>
<th style="text-align:left;">
Rat_Options
</th>
<th style="text-align:left;">
Drosophila_Options
</th>
<th style="text-align:left;">
Macaque_Options
</th>
<th style="text-align:left;">
Chicken_Options
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
Mouse
</td>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
Marmoset
</td>
<td style="text-align:left;">
Zebrafish
</td>
<td style="text-align:left;">
Rat
</td>
<td style="text-align:left;">
Drosophila
</td>
<td style="text-align:left;">
Macaque
</td>
<td style="text-align:left;">
Chicken
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
mouse
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:left;">
marmoset
</td>
<td style="text-align:left;">
zebrafish
</td>
<td style="text-align:left;">
rat
</td>
<td style="text-align:left;">
drosophila
</td>
<td style="text-align:left;">
macaque
</td>
<td style="text-align:left;">
chicken
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
Ms
</td>
<td style="text-align:left;">
Hu
</td>
<td style="text-align:left;">
CJ
</td>
<td style="text-align:left;">
DR
</td>
<td style="text-align:left;">
RN
</td>
<td style="text-align:left;">
DM
</td>
<td style="text-align:left;">
Rhesus
</td>
<td style="text-align:left;">
Gallus
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
ms
</td>
<td style="text-align:left;">
hu
</td>
<td style="text-align:left;">
Cj
</td>
<td style="text-align:left;">
Dr
</td>
<td style="text-align:left;">
Rn
</td>
<td style="text-align:left;">
Dm
</td>
<td style="text-align:left;">
macaca
</td>
<td style="text-align:left;">
gallus
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
Mm
</td>
<td style="text-align:left;">
Hs
</td>
<td style="text-align:left;">
cj
</td>
<td style="text-align:left;">
dr
</td>
<td style="text-align:left;">
rn
</td>
<td style="text-align:left;">
dm
</td>
<td style="text-align:left;">
mmulatta
</td>
<td style="text-align:left;">
Gg
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
mm
</td>
<td style="text-align:left;">
hs
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
gg
</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="non-default-species">Non-default species<a class="anchor" aria-label="anchor" href="#non-default-species"></a>
</h5>
<p>However custom prefixes can be used for species with different
annotations. Simply specify <code>species = other</code> and supply
feature lists or regex patterns for your species of interest. <em>NOTE:
If desired please submit issue on GitHub for additional default species.
Please include regex pattern or list of genes for both mitochondrial and
ribosomal genes and I will add additional built-in defaults to the
function.</em></p>
<details><summary><strong>What is example of how this works?</strong>
</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using gene name patterns</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"regexp_pattern_mito"</span>, ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern_ribo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using feature name lists</span></span>
<span><span class="va">mito_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span></span>
<span><span class="va">ribo_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>, ribo_features <span class="op">=</span> <span class="va">ribo_gene_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using combination of gene lists and gene name patterns</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"Human"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>, ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern_ribo"</span><span class="op">)</span></span></code></pre></div>
</details><p> </p>
</div>
</div>
<div class="section level3">
<h3 id="add-all-cell-qc-metrics-with-single-function">Add All Cell QC Metrics with Single Function<a class="anchor" aria-label="anchor" href="#add-all-cell-qc-metrics-with-single-function"></a>
</h3>
<p>To simplify the process of adding cell QC metrics scCustomize
contains a wrapper function which can be customized to add all or some
of the available QC metrics. This vignette will describe each of these
in more detail below but using the default parameters of the function
<code><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics()</a></code> will add:</p>
<ul>
<li>Mitochondrial and Ribosomal Percentages (default and custom
species).</li>
<li>Hemoglobin percentages (default and custom species).<br>
</li>
<li>Cell Complexity (log10(nFeature) / log10(nCount).</li>
<li>Top XX Gene Percentage.</li>
<li>Percentage of counts for IEG (human and mouse only).</li>
<li>OXPHOS, APOP, and DNA Repair pathways (supported species only).</li>
<li>Cell Cycle Scoring (Human only).</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Mito/Ribo Percentages</span> to meta.data.</span></span>
<span><span class="co">## Adding Percent Mitochondrial genes for <span style="color: #00BB00;">human</span> using gene symbol pattern: <span style="color: #0000BB;">"^MT-"</span>.</span></span>
<span><span class="co">## Adding Percent Ribosomal genes for <span style="color: #00BB00;">human</span> using gene symbol pattern: <span style="color: #0000BB;">"^RP[SL]"</span>.</span></span>
<span><span class="co">## Adding Percent Mito+Ribo by adding Mito &amp; Ribo percentages.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Cell Complexity #1 (log10GenesPerUMI)</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Cell Complexity #2 (Top 50 Percentages)</span> to meta.data.</span></span>
<span><span class="co">## Calculating percent expressing top 50 for layer: <span style="color: #00BB00;">counts</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">MSigDB Oxidative Phosphorylation, Apoptosis, and DNA Repair</span></span></span>
<span><span class="co">##   <span style="color: #00BB00;">Percentages</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">IEG Percentages</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Hemoglobin Percentages</span> to meta.data.</span></span>
<span><span class="co">## Adding Percent Hemoglobin for <span style="color: #00BB00;">Human</span> using gene symbol pattern: <span style="color: #0000BB;">"^HB[^(P)]"</span>.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Cell Cycle Scoring</span> to meta.data.</span></span>
<span><span class="co">## Calculating <span style="color: #00BB00;">Cell Cycle Scores</span>.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="add-qc-metrics-individually">Add QC Metrics Individually<a class="anchor" aria-label="anchor" href="#add-qc-metrics-individually"></a>
</h3>
<p>If you only want to add some but not all metrics you can either
customize <code>Add_Cell_QC_Metrics</code> or use the individual
functions.</p>
<div class="section level4">
<h4 id="add-mitochondrial-ribosomal-percentages">Add Mitochondrial &amp; Ribosomal Percentages<a class="anchor" aria-label="anchor" href="#add-mitochondrial-ribosomal-percentages"></a>
</h4>
<p>If you just want to calculate and add mitochondrial and ribosomal
count percentages per cell/nucleus you can use
<code>Add_Mito_Ribo</code>.</p>
<div class="section level5">
<h5 id="add_mito_ribo">
<code>Add_Mito_Ribo()</code><a class="anchor" aria-label="anchor" href="#add_mito_ribo"></a>
</h5>
<p>scCustomize contains easy wrapper function to automatically add both
Mitochondrial and Ribosomal percentages to meta.data slot. If you are
using mouse, human, marmoset, zebrafish, rat, drosophila, rhesus
macaque, or chicken data all you need to do is specify the
<code>species</code> parameter.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These defaults can be run just by providing accepted species name</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="analysis-with-two-species">Analysis with two species<a class="anchor" aria-label="anchor" href="#analysis-with-two-species"></a>
</h5>
<p>Some analyses are performed with cells aligned to a genome that
contains multiple species (see Cell Ranger/10X documentation for more
info). scCustomize now supports adding mitochondrial and ribosomal
percentages for these datasets using optional parameters. Here we will
use example data provided by 10X Genomics <a href="https://www.10xgenomics.com/datasets/10k-hgmm-3p-gemx" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dual_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"~/Downloads/10k_hgmm_3p_gemx_Multiplex_count_raw_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_dual_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">pbmc_dual_species</span>, min.cells <span class="op">=</span> <span class="fl">5</span>, min.features <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<p>For dual species analyses the only other information you need to
provide is what the prefixes are used in front of gene IDs. In this case
the prefixes are “GRCh38-” and “GRCm39-”.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dual_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_dual_species</span>, species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"mouse"</span><span class="op">)</span>, species_prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GRCh38-"</span>,</span>
<span>    <span class="st">"GRCm39-"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="warning-messages">Warning Messages<a class="anchor" aria-label="anchor" href="#warning-messages"></a>
</h5>
<p>The added benefit of <code>Add_Mito_Ribo</code> is that it will
return informative warnings if no Mitochondrial or Ribosomal features
are found using the current species, features, or pattern
specification.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For demonstration purposes we can set `species = mouse` for this object of human cells</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `Add_Mito_Ribo()`:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> Columns with <span style="color: #0000BB;">"percent_mito"</span> and/or <span style="color: #0000BB;">"percent_ribo"</span> already present in</span></span>
<span><span class="co">##   meta.data slot.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> *To run function and overwrite columns set parameter `overwrite = TRUE` or</span></span>
<span><span class="co">##   change respective `mito_name`, `ribo_name`, and/or `mito_ribo_name`*</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or if providing custom patterns/lists and features not found</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"^MT-"</span>, ribo_pattern <span class="op">=</span> <span class="st">"BAD_PATTERN"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: No Ribo features found in object using pattern/feature list provided.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> No column will be added to meta.data.</span></span></code></pre>
<pre><code><span><span class="co">## Adding Percent Mitochondrial genes for <span style="color: #00BB00;">other</span> using gene symbol</span></span>
<span><span class="co">## pattern: <span style="color: #0000BB;">"^MT-"</span>.</span></span></code></pre>
<p><code>Add_Mito_Ribo</code> will also return warnings if columns are
already present in <code>@meta.data</code> slot and prompt you to
provide override if you want to run the function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `Add_Mito_Ribo()`:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> Columns with <span style="color: #0000BB;">"percent_mito"</span> and/or <span style="color: #0000BB;">"percent_ribo"</span> already present in</span></span>
<span><span class="co">##   meta.data slot.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> *To run function and overwrite columns set parameter `overwrite = TRUE` or</span></span>
<span><span class="co">##   change respective `mito_name`, `ribo_name`, and/or `mito_ribo_name`*</span></span></code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="add-cell-complexitynovelty-qc-metrics">Add Cell Complexity/Novelty QC Metrics<a class="anchor" aria-label="anchor" href="#add-cell-complexitynovelty-qc-metrics"></a>
</h3>
<p>In addition to metrics like number of features and UMIs it can often
be helpful to analyze the complexity of expression within a single cell.
scCustomize provides functions to add two of these metrics to meta
data.</p>
<div class="section level4">
<h4 id="cell-complexity-log10nfeature-log10ncount">Cell Complexity (log10(nFeature) / log10(nCount))<a class="anchor" aria-label="anchor" href="#cell-complexity-log10nfeature-log10ncount"></a>
</h4>
<p>scCustomize contains easy shortcut function to add a measure of cell
complexity/novelty that can sometimes be useful to filter low quality
cells. The metric is calculated by calculating the result of
log10(nFeature) / log10(nCount).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These defaults can be run just by providing accepted species name</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_Complexity.html">Add_Cell_Complexity</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="add-top-percent-expression-qc-metric">Add Top Percent Expression QC Metric<a class="anchor" aria-label="anchor" href="#add-top-percent-expression-qc-metric"></a>
</h4>
<p>Additionally, (or alternatively), scCustomize contains another metric
of complexity which is the top percent expression. The user supplies an
integer value for <code>num_top_genes</code> (default is 50) which
species the number of genes and the function returns percentage of
counts occupied by top XX genes in each cell.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These defaults can be run just by providing accepted species name</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Top_Gene_Pct.html">Add_Top_Gene_Pct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, num_top_genes <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="add-hemoglobin-percentage">Add Hemoglobin Percentage<a class="anchor" aria-label="anchor" href="#add-hemoglobin-percentage"></a>
</h3>
<p>scCustomize also contains function to add percentage of counts for
hemoglobin genes. Use of this metric is much more situational. If your
experiment has the potential for red blood cell contamination but you
want to avoid that then this can be helpful. A high percentage of
hemoglobin counts may indicate that your sample has high amount of
ambient RNA present or RBCs in the cells captured.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Hemo.html">Add_Hemo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-qc-metrics-from-pathway-gene-lists">Add QC Metrics from Pathway Gene Lists<a class="anchor" aria-label="anchor" href="#add-qc-metrics-from-pathway-gene-lists"></a>
</h3>
<p>In addition to those standard QC metrics it can be helpful when using
networ- based QC analysis to add the percent of expression of genes
related to common pathways. This function and the network-based analysis
is further extension of the analysis/QC from our recent publication:
Gazestani &amp; Kamath et al., 2023 (<a href="https://doi.org/10.1016/j.cell.2023.08.005" class="external-link"><em>Cell</em></a>).</p>
<p>In scCustomize the percent of gene expression from the following gene
lists can be added as part of the <code>Add_Cell_QC_Metrics</code>:</p>
<ul>
<li>Immediate Early Genes (for human and mouse only)
<ul>
<li>Can be used in part to examine potential impact of dissociation or
post-mortem signatures (<a href="https://doi.org/10.1038/s41593-022-01022-8" class="external-link"><em>Marsh et al.,
2022</em></a>) or to identify acutely perturbed populations (gene list
from <a href="https://doi.org/10.1016/j.neuron.2017.09.026" class="external-link"><em>Wu et
al., 2017</em></a>)</li>
</ul>
</li>
<li>Oxidative Phosphorylation, Apoptosis, &amp; DNA Repair (all default
species except Marmoset)
<ul>
<li>Species specific gene lists from <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=H" class="external-link">MSigDB
Hallmark Gene Sets</a>
</li>
</ul>
</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
