<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Object QC Functions • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Object QC Functions">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
    <li>
      <a href="../articles/cNMF_Functionality.html">cNMF Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Object_QC_Functions_files/kePrint-0.0.1/kePrint.js"></script><link href="Object_QC_Functions_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Object QC Functions</h1>
            
            <h4 data-toc-skip class="date">Compiled: September 10,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/articles/Object_QC_Functions.Rmd" class="external-link"><code>vignettes/articles/Object_QC_Functions.Rmd</code></a></small>
      <div class="hidden name"><code>Object_QC_Functions.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="quality-control-of-scrna-seq-objects">Quality control of scRNA-seq objects<a class="anchor" aria-label="anchor" href="#quality-control-of-scrna-seq-objects"></a>
</h2>
<p>scCustomize has several helper functions to simplify/streamline what
is nearly always the first and most critical choices when starting an
analysis: performing quality control and filtering.</p>
<p>Let’s load packages and raw data object for this tutorial.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">pbmc3k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/pbmc3k.SeuratData/man/pbmc3k.html" class="external-link">pbmc3k</a></span></span>
<span><span class="va">hcabm</span> <span class="op">&lt;-</span> <span class="fu">hcabm40k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/hcabm40k.SeuratData/man/hcabm40k.html" class="external-link">hcabm40k</a></span></span></code></pre></div>
<p>We’ll add some random meta data variables to pbmc data form use in
this vignette</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span>, <span class="st">"sample4"</span>, <span class="st">"sample5"</span>, <span class="st">"sample6"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,</span>
<span>    replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Batch1"</span>, <span class="st">"Batch2"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="cross-compatibility-of-qc-functions">Cross-compatibility of QC Functions<a class="anchor" aria-label="anchor" href="#cross-compatibility-of-qc-functions"></a>
</h3>
<div class="section level4">
<h4 id="support-for-seurat-and-liger-objects">Support for Seurat and LIGER Objects<a class="anchor" aria-label="anchor" href="#support-for-seurat-and-liger-objects"></a>
</h4>
<p>All of scCustomize’s functions to add quality control metrics are
100% cross compatible across Seurat and LIGER objects using the same
function calls. For more details on QC-specific plotting functions see
<a href="https://samuel-marsh.github.io/scCustomize/articles/QC_Plots.html">QC
Plotting &amp; Analysis Vignette</a>.</p>
</div>
<div class="section level4">
<h4 id="support-for-gene-symbols-and-ensembl-ids">Support for gene symbols and Ensembl IDs<a class="anchor" aria-label="anchor" href="#support-for-gene-symbols-and-ensembl-ids"></a>
</h4>
<p>Additionally, all of the QC functions support objects that use either
gene symbols or Ensembl IDs. Ensembl IDs for default species (see below)
are from Ensembl version 112 (updated in scCustomize on 4/29/2024).</p>
<p>If your object using ensembl IDs as features names then simply add
<code>ensembl_ids</code> parameter that is present in all QC
functions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using gene name patterns</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, species <span class="op">=</span> <span class="st">"Human"</span>, ensembl_ids <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="default-species-support">Default Species Support<a class="anchor" aria-label="anchor" href="#default-species-support"></a>
</h3>
<p>Many of the QC functions commonly performed depend on genes within a
particular family that have similar naming patterns (e.g., Mitochondrial
genes) or are species specific (see msigdb dependent parts of
<code><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics()</a></code>).</p>
<p>To simplify the process of needing to remember species-specific
patterns (or find Ensembl ID gene lists)</p>
<p>If you are using mouse, human, marmoset, zebrafish, rat, drosophila,
rhesus macaque, or chicken data all you need to do is specify the
<code>species</code> parameter in the functions described below using
one of the following accepted names.</p>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Mouse_Options
</th>
<th style="text-align:left;">
Human_Options
</th>
<th style="text-align:left;">
Marmoset_Options
</th>
<th style="text-align:left;">
Zebrafish_Options
</th>
<th style="text-align:left;">
Rat_Options
</th>
<th style="text-align:left;">
Drosophila_Options
</th>
<th style="text-align:left;">
Macaque_Options
</th>
<th style="text-align:left;">
Chicken_Options
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
Mouse
</td>
<td style="text-align:left;">
Human
</td>
<td style="text-align:left;">
Marmoset
</td>
<td style="text-align:left;">
Zebrafish
</td>
<td style="text-align:left;">
Rat
</td>
<td style="text-align:left;">
Drosophila
</td>
<td style="text-align:left;">
Macaque
</td>
<td style="text-align:left;">
Chicken
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
mouse
</td>
<td style="text-align:left;">
human
</td>
<td style="text-align:left;">
marmoset
</td>
<td style="text-align:left;">
zebrafish
</td>
<td style="text-align:left;">
rat
</td>
<td style="text-align:left;">
drosophila
</td>
<td style="text-align:left;">
macaque
</td>
<td style="text-align:left;">
chicken
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
Ms
</td>
<td style="text-align:left;">
Hu
</td>
<td style="text-align:left;">
CJ
</td>
<td style="text-align:left;">
DR
</td>
<td style="text-align:left;">
RN
</td>
<td style="text-align:left;">
DM
</td>
<td style="text-align:left;">
Rhesus
</td>
<td style="text-align:left;">
Gallus
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
ms
</td>
<td style="text-align:left;">
hu
</td>
<td style="text-align:left;">
Cj
</td>
<td style="text-align:left;">
Dr
</td>
<td style="text-align:left;">
Rn
</td>
<td style="text-align:left;">
Dm
</td>
<td style="text-align:left;">
macaca
</td>
<td style="text-align:left;">
gallus
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
Mm
</td>
<td style="text-align:left;">
Hs
</td>
<td style="text-align:left;">
cj
</td>
<td style="text-align:left;">
dr
</td>
<td style="text-align:left;">
rn
</td>
<td style="text-align:left;">
dm
</td>
<td style="text-align:left;">
mmulatta
</td>
<td style="text-align:left;">
Gg
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
mm
</td>
<td style="text-align:left;">
hs
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
gg
</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="non-default-species">Non-default species<a class="anchor" aria-label="anchor" href="#non-default-species"></a>
</h5>
<p>However custom prefixes can be used for species with different
annotations. Simply specify <code>species = other</code> and supply
feature lists or regex patterns for your species of interest. <em>NOTE:
If desired please submit issue on GitHub for additional default species.
Please include regex pattern or list of genes for both mitochondrial and
ribosomal genes and I will add additional built-in defaults to the
function.</em></p>
<details><summary><strong>What is example of how this works?</strong>
</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using gene name patterns</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"regexp_pattern_mito"</span>,</span>
<span>    ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern_ribo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using feature name lists</span></span>
<span><span class="va">mito_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span></span>
<span><span class="va">ribo_gene_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene1"</span>, <span class="st">"gene2"</span>, <span class="st">"etc"</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>, ribo_features <span class="op">=</span> <span class="va">ribo_gene_list</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using combination of gene lists and gene name patterns</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_features <span class="op">=</span> <span class="va">mito_gene_list</span>, ribo_pattern <span class="op">=</span> <span class="st">"regexp_pattern_ribo"</span><span class="op">)</span></span></code></pre></div>
</details><p> </p>
</div>
</div>
<div class="section level3">
<h3 id="add-all-cell-qc-metrics-with-single-function">Add All Cell QC Metrics with Single Function<a class="anchor" aria-label="anchor" href="#add-all-cell-qc-metrics-with-single-function"></a>
</h3>
<p>To simplify the process of adding cell QC metrics scCustomize
contains a wrapper function which can be customized to add all or some
of the available QC metrics. This vignette will describe each of these
in more detail below but using the default parameters of the function
<code><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics()</a></code> will add:</p>
<ul>
<li>Mitochondrial and Ribosomal Percentages (default and custom
species).</li>
<li>Hemoglobin percentages (default and custom species).<br>
</li>
<li>Percentage of counts for lncRNA (default species).<br>
</li>
<li>Cell Complexity (log10(nFeature) / log10(nCount).</li>
<li>Top XX Gene Percentage.</li>
<li>Percentage of counts for IEG (human and mouse only).</li>
<li>Module score for IEG genes (human and mouse only).<br>
</li>
<li>OXPHOS, APOP, and DNA Repair pathways (supported species only).</li>
<li>Cell Cycle Scoring (Human only).</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Mito/Ribo Percentages</span> to meta.data.</span></span>
<span><span class="co">## Adding Percent Mitochondrial genes for <span style="color: #00BB00;">human</span> using gene symbol pattern: <span style="color: #0000BB;">"^MT-"</span>.</span></span>
<span><span class="co">## Adding Percent Ribosomal genes for <span style="color: #00BB00;">human</span> using gene symbol pattern: <span style="color: #0000BB;">"^RP[SL]"</span>.</span></span>
<span><span class="co">## Adding Percent Mito+Ribo by adding Mito &amp; Ribo percentages.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Cell Complexity #1 (log10GenesPerUMI)</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Cell Complexity #2 (Top 50 Percentages)</span> to meta.data.</span></span>
<span><span class="co">## Calculating percent expressing top 50 for layer: <span style="color: #00BB00;">counts</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">MSigDB Oxidative Phosphorylation, Apoptosis, and DNA Repair</span></span></span>
<span><span class="co">##   <span style="color: #00BB00;">Percentages</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">IEG Percentages</span> &amp; <span style="color: #00BB00;">IEG Module Score</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Hemoglobin Percentages</span> to meta.data.</span></span>
<span><span class="co">## Adding Percent Hemoglobin for <span style="color: #00BB00;">Human</span> using gene symbol pattern: <span style="color: #0000BB;">"^HB[^(P)]"</span>.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">lncRNA Percentages</span> to meta.data.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">•</span> Adding <span style="color: #00BB00;">Cell Cycle Scoring</span> to meta.data.</span></span>
<span><span class="co">## Calculating <span style="color: #00BB00;">Cell Cycle Scores</span>.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="add-qc-metrics-individually">Add QC Metrics Individually<a class="anchor" aria-label="anchor" href="#add-qc-metrics-individually"></a>
</h3>
<p>If you only want to add some but not all metrics you can either
customize <code>Add_Cell_QC_Metrics</code> or use the individual
functions.</p>
<div class="section level4">
<h4 id="add-mitochondrial-ribosomal-percentages">Add Mitochondrial &amp; Ribosomal Percentages<a class="anchor" aria-label="anchor" href="#add-mitochondrial-ribosomal-percentages"></a>
</h4>
<p>If you just want to calculate and add mitochondrial and ribosomal
count percentages per cell/nucleus you can use
<code>Add_Mito_Ribo</code>.</p>
<div class="section level5">
<h5 id="add_mito_ribo">
<code>Add_Mito_Ribo()</code><a class="anchor" aria-label="anchor" href="#add_mito_ribo"></a>
</h5>
<p>scCustomize contains easy wrapper function to automatically add both
Mitochondrial and Ribosomal percentages to meta.data slot. If you are
using mouse, human, marmoset, zebrafish, rat, drosophila, rhesus
macaque, or chicken data all you need to do is specify the
<code>species</code> parameter.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These defaults can be run just by providing accepted species name</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="analysis-with-two-species">Analysis with two species<a class="anchor" aria-label="anchor" href="#analysis-with-two-species"></a>
</h5>
<p>Some analyses are performed with cells aligned to a genome that
contains multiple species (see Cell Ranger/10X documentation for more
info). scCustomize now supports adding mitochondrial and ribosomal
percentages for these datasets using optional parameters. Here we will
use example data provided by 10X Genomics <a href="https://www.10xgenomics.com/datasets/10k-hgmm-3p-gemx" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dual_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">"~/Downloads/10k_hgmm_3p_gemx_Multiplex_count_raw_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_dual_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">pbmc_dual_species</span>, min.cells <span class="op">=</span> <span class="fl">5</span>, min.features <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<p>For dual species analyses the only other information you need to
provide is what the prefixes are used in front of gene IDs. In this case
the prefixes are “GRCh38-” and “GRCm39-”.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_dual_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc_dual_species</span>, species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"mouse"</span><span class="op">)</span>, species_prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GRCh38-"</span>,</span>
<span>    <span class="st">"GRCm39-"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="warning-messages">Warning Messages<a class="anchor" aria-label="anchor" href="#warning-messages"></a>
</h5>
<p>The added benefit of <code>Add_Mito_Ribo</code> is that it will
return informative warnings if no Mitochondrial or Ribosomal features
are found using the current species, features, or pattern
specification.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For demonstration purposes we can set `species = mouse` for this object of human cells</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `Add_Mito_Ribo()`:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> Columns with <span style="color: #0000BB;">"percent_mito"</span> and/or <span style="color: #0000BB;">"percent_ribo"</span> already present in</span></span>
<span><span class="co">##   meta.data slot.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> *To run function and overwrite columns set parameter `overwrite = TRUE` or</span></span>
<span><span class="co">##   change respective `mito_name`, `ribo_name`, and/or `mito_ribo_name`*</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Or if providing custom patterns/lists and features not found</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"other"</span>, mito_pattern <span class="op">=</span> <span class="st">"^MT-"</span>, ribo_pattern <span class="op">=</span> <span class="st">"BAD_PATTERN"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: No Ribo features found in object using pattern/feature list provided.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> No column will be added to meta.data.</span></span></code></pre>
<pre><code><span><span class="co">## Adding Percent Mitochondrial genes for <span style="color: #00BB00;">other</span> using gene symbol</span></span>
<span><span class="co">## pattern: <span style="color: #0000BB;">"^MT-"</span>.</span></span></code></pre>
<p><code>Add_Mito_Ribo</code> will also return warnings if columns are
already present in <code>@meta.data</code> slot and prompt you to
provide override if you want to run the function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `Add_Mito_Ribo()`:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> Columns with <span style="color: #0000BB;">"percent_mito"</span> and/or <span style="color: #0000BB;">"percent_ribo"</span> already present in</span></span>
<span><span class="co">##   meta.data slot.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> *To run function and overwrite columns set parameter `overwrite = TRUE` or</span></span>
<span><span class="co">##   change respective `mito_name`, `ribo_name`, and/or `mito_ribo_name`*</span></span></code></pre>
</div>
</div>
</div>
<div class="section level3">
<h3 id="add-cell-complexitynovelty-qc-metrics">Add Cell Complexity/Novelty QC Metrics<a class="anchor" aria-label="anchor" href="#add-cell-complexitynovelty-qc-metrics"></a>
</h3>
<p>In addition to metrics like number of features and UMIs it can often
be helpful to analyze the complexity of expression within a single cell.
scCustomize provides functions to add two of these metrics to meta
data.</p>
<div class="section level4">
<h4 id="cell-complexity-log10nfeature-log10ncount">Cell Complexity (log10(nFeature) / log10(nCount))<a class="anchor" aria-label="anchor" href="#cell-complexity-log10nfeature-log10ncount"></a>
</h4>
<p>scCustomize contains easy shortcut function to add a measure of cell
complexity/novelty that can sometimes be useful to filter low quality
cells. The metric is calculated by calculating the result of
log10(nFeature) / log10(nCount).</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These defaults can be run just by providing accepted species name</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_Complexity.html">Add_Cell_Complexity</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="add-top-percent-expression-qc-metric">Add Top Percent Expression QC Metric<a class="anchor" aria-label="anchor" href="#add-top-percent-expression-qc-metric"></a>
</h4>
<p>Additionally, (or alternatively), scCustomize contains another metric
of complexity which is the top percent expression. The user supplies an
integer value for <code>num_top_genes</code> (default is 50) which
species the number of genes and the function returns percentage of
counts occupied by top XX genes in each cell.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># These defaults can be run just by providing accepted species name</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Top_Gene_Pct.html">Add_Top_Gene_Pct</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, num_top_genes <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="add-hemoglobin-percentage">Add Hemoglobin Percentage<a class="anchor" aria-label="anchor" href="#add-hemoglobin-percentage"></a>
</h3>
<p>scCustomize also contains function to add percentage of counts for
hemoglobin genes. Use of this metric is much more situational. If your
experiment has the potential for red blood cell contamination but you
want to avoid that then this can be helpful. A high percentage of
hemoglobin counts may indicate that your sample has high amount of
ambient RNA present or RBCs in the cells captured.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Hemo.html">Add_Hemo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-qc-metrics-from-pathway-gene-lists">Add QC Metrics from Pathway Gene Lists<a class="anchor" aria-label="anchor" href="#add-qc-metrics-from-pathway-gene-lists"></a>
</h3>
<p>In addition to those standard QC metrics it can be helpful when using
networ- based QC analysis to add the percent of expression of genes
related to common pathways. This function and the network-based analysis
is further extension of the analysis/QC from our recent publication:
Gazestani &amp; Kamath et al., 2023 (<a href="https://doi.org/10.1016/j.cell.2023.08.005" class="external-link"><em>Cell</em></a>).</p>
<p>In scCustomize the percent of gene expression from the following gene
lists can be added as part of the <code>Add_Cell_QC_Metrics</code>:</p>
<ul>
<li>Immediate Early Genes (for human and mouse only)
<ul>
<li>Can be used in part to examine potential impact of dissociation or
post-mortem signatures (<a href="https://doi.org/10.1038/s41593-022-01022-8" class="external-link"><em>Marsh et al.,
2022</em></a>) or to identify acutely perturbed populations (gene list
from <a href="https://doi.org/10.1016/j.neuron.2017.09.026" class="external-link"><em>Wu et
al., 2017</em></a>)</li>
</ul>
</li>
<li>Oxidative Phosphorylation, Apoptosis, &amp; DNA Repair (all default
species except Marmoset)
<ul>
<li>Species specific gene lists from <a href="https://www.gsea-msigdb.org/gsea/msigdb/human/genesets.jsp?collection=H" class="external-link">MSigDB
Hallmark Gene Sets</a>
</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="calculate-and-add-malat1-threshold">Calculate and Add MALAT1 Threshold<a class="anchor" aria-label="anchor" href="#calculate-and-add-malat1-threshold"></a>
</h3>
<p>Another QC metric that has been proposed is based on the level of
expression of long non-coding RNA MALAT1 (<a href="https://doi.org/10.1101/2024.07.14.603469%7D" class="external-link"><em>Clarke &amp;
Bader, 2024</em></a>). Their analysis demonstrates that low MALAT1
expression is characteristic of low quality cells and created function
to determine a cutoff threshold to call low quality/damaged cells. Their
metric has now been incorporated into scCustomize as
<code>Add_MALAT1_Threshold</code>.</p>
<p><strong>NOTE: If you utilize this function please cite the above
manuscript in addition to scCustomize).</strong></p>
<p><strong>NOTE: This function is currently only available for human and
mouse as those are only species with MALAT1 annotated. See section below
on non-default species for more info.</strong></p>
<p>For best results this function should be run on per sample basis
using normalized count matrix. To get started all the function requires
is Seurat Object, the species being analyzed, and the name of meta.data
column that contains sample ID information.</p>
<p>For this example we are going to use a subset of data from Hammond et
al., 2019.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hammond_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_MALAT1_Threshold.html">Add_MALAT1_Threshold</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hammond_subset</span>, species <span class="op">=</span> <span class="st">"mouse"</span>, sample_col <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">•</span> Adding MALAT1 Threshold for <span style="color: #00BB00;">mouse</span> using gene id: <span style="color: #0000BB;">"Malat1"</span>.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> <span style="color: #00BBBB;">Please cite</span> <span style="color: #00BB00;">Clarke &amp; Bader (2024). doi.org/10.1101/2024.07.14.603469</span> <span style="color: #00BBBB;">when</span></span></span>
<span><span class="co">##   <span style="color: #00BBBB;">using MALAT1 thresholding function.</span></span></span>
<span><span class="co">## Calculating thresholds across <span style="color: #00BB00;">4</span> samples from meta.data column <span style="color: #00BB00;">orig.ident</span>.</span></span>
<span><span class="co">## A total of <span style="color: #00BB00;">280</span> cells (<span style="color: #00BB00;">3%</span>) fall below the threshold.</span></span>
<span><span class="co">## Adding results to object as <span style="color: #0000BB;">"Malat1_Threshold"</span>.</span></span></code></pre>
<p><strong>NOTE: The function is also compatible with Ensembl IDs by
specifying <code>ensembl_ids = TRUE</code> in function
call.</strong></p>
<div class="section level5">
<h5 id="single-sample-object">Single Sample Object<a class="anchor" aria-label="anchor" href="#single-sample-object"></a>
</h5>
<p>If your Seurat object only has a single sample or you want to run the
calculation across the entire object then you can supply the optional
parameter of <code>whole_object = TRUE</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hcabm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_MALAT1_Threshold.html">Add_MALAT1_Threshold</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hammond_subset</span>, species <span class="op">=</span> <span class="st">"mouse"</span>, whole_object <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="non-default-species-1">Non-default species<a class="anchor" aria-label="anchor" href="#non-default-species-1"></a>
</h5>
<p>Currently, humans and mice are the only species for which MALAT1 has
been annotated in the genome. If your analysis includes annotation of
MALAT1 homolog/paralog in new species you can supply that value to
optional parameter <code>homolog_name</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hcabm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_MALAT1_Threshold.html">Add_MALAT1_Threshold</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">OBJECT</span>, species <span class="op">=</span> <span class="st">"marmoset"</span>, homolog_name <span class="op">=</span> <span class="st">"MALAT1_HOMOLOG_SYMBOL_OR_ID_NUMBER"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualizing-the-results">Visualizing the Results<a class="anchor" aria-label="anchor" href="#visualizing-the-results"></a>
</h4>
<p>The function will calculate the threshold for each sample and adds
new column to meta.data containing logical value (TRUE/FALSE)
designating cells above or below the threshold. This can then be
visualized as any meta.data column.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot_scCustom.html">VlnPlot_scCustom</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hammond_subset</span>, features <span class="op">=</span> <span class="st">"nFeature_RNA"</span>, split.by <span class="op">=</span> <span class="st">"Malat1_Threshold"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Object_QC_Functions_files/figure-html/unnamed-chunk-23-1.png" width="1248" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/FeatureScatter_scCustom.html">FeatureScatter_scCustom</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hammond_subset</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span>,</span>
<span>    split.by <span class="op">=</span> <span class="st">"Malat1_Threshold"</span>, pt.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Object_QC_Functions_files/figure-html/unnamed-chunk-23-2.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="visualizing-saving-threshold-calculation-plots">Visualizing &amp; Saving Threshold Calculation Plots<a class="anchor" aria-label="anchor" href="#visualizing-saving-threshold-calculation-plots"></a>
</h4>
<p>In addition to basic parameters there are several optional parameters
that can be invoked. The first set relate to a series of plots the are
created when determining the threshold. For detailed description of
these plots and their interpretation please see <a href="https://github.com/BaderLab/MALAT1_threshold" class="external-link">Clarke &amp; Bader
GitHub Repo</a>. The optional parameters within scCustomize are:<br>
- <code>print_plots</code> whether or not to print plots to output while
running function (default is FALSE).<br>
- <code>save_plots</code> whether or not to save plots to PDF file while
running function (default is FALSE).<br>
- <code>save_plot_path</code> directory path to save the plots.<br>
- <code>save_plot_name</code> file name for PDF document with
plots.<br>
- <code>plot_width</code> plot width (inches) - <code>plot_height</code>
plot height (inches)</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hcabm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_MALAT1_Threshold.html">Add_MALAT1_Threshold</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hammond_subset</span>, species <span class="op">=</span> <span class="st">"mouse"</span>, sample_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>    save_plots <span class="op">=</span> <span class="cn">TRUE</span>, save_plot_path <span class="op">=</span> <span class="st">"plots/QC_Plots/"</span>, save_plot_name <span class="op">=</span> <span class="st">"MALAT1_Threshold_Plots"</span><span class="op">)</span></span></code></pre></div>
Here is eample of plots for one of the samples in the Hammond et al.,
subset dataset.
<center>
<img src="../reference/figures/assets/malat1_example.png" width="49%"><br><em>Example outputs with default naming based on file name or providing
<code>sample_names</code> parameter.</em>
</center>
</div>
<div class="section level4">
<h4 id="optional-parameters-to-modify-threshold-calculation">Optional Parameters to Modify Threshold Calculation<a class="anchor" aria-label="anchor" href="#optional-parameters-to-modify-threshold-calculation"></a>
</h4>
<p>Finally, there are also a series of parameters of the MALAT1 function
itself that can be modified. For deeper description of parameters below
please see manual documentation for <code><a href="../reference/Add_MALAT1_Threshold.html">Add_MALAT1_Threshold()</a></code>
and above mentioned manuscript: <code>bw</code>, <code>lwd</code>,
<code>breaks</code>, <code>chosen_min</code>, <code>smooth</code>,
<code>abs_min</code>, <code>rough_max</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="add-exam-module-score">Add exAM Module Score<a class="anchor" aria-label="anchor" href="#add-exam-module-score"></a>
</h3>
<p>In Marsh et al., 2022 (<a href="https://doi.org/10.1038/s41593-022-01022-8" class="external-link"><em>Nature
Neuroscience</em></a>) we identified sets of genes that were
differentially expressed in samples that were processed with enzymatic
digestion or in post-mortem human samples compared to fresh tissue (see
paper for full details). Module scores for these gene sets can be added
to Seurat object using function <code>exAM_Scoring</code>.</p>
<p>To demonstrate this function we will use data objects from Marsh et
al., 2022.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Marsh et al., 2022 datasets</span></span>
<span><span class="va">marsh_mouse_micro</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"assets/marsh_2020_micro.qs"</span><span class="op">)</span></span>
<span><span class="va">marsh_human_pm</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"assets/marsh_human_pm.qs"</span><span class="op">)</span></span></code></pre></div>
<p>If adding score to human data both the union score and “microglia”
(post-mortem) score will be added to be meta.data slot. When species is
mouse only the union score will be added. When the function is run the
name of the column added to meta.data slot will be included printed in
message.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marsh_mouse_micro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exAM_Scoring.html">exAM_Scoring</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">marsh_mouse_micro</span>, species <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> <span style="color: #00BBBB;">Please cite</span> <span style="color: #00BB00;">Marsh et al., (2022). doi.org/10.1038/s41593-022-01022-8</span> <span style="color: #00BBBB;">when</span></span></span>
<span><span class="co">##   <span style="color: #00BBBB;">using exAM Scoring function.</span></span></span>
<span><span class="co">## Adding module score for exAM union gene list as <span style="color: #00BB00;">“exAM_Union_Score”</span>.</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marsh_human_pm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exAM_Scoring.html">exAM_Scoring</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">marsh_human_pm</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> <span style="color: #00BBBB;">Please cite</span> <span style="color: #00BB00;">Marsh et al., (2022). doi.org/10.1038/s41593-022-01022-8</span> <span style="color: #00BBBB;">when</span></span></span>
<span><span class="co">##   <span style="color: #00BBBB;">using exAM Scoring function.</span></span></span>
<span><span class="co">## Adding module score for exAM union gene list as <span style="color: #00BB00;">“exAM_Union_Score”</span>.</span></span>
<span><span class="co">## Adding module score for exAM Microglia Factor gene list as</span></span>
<span><span class="co">## <span style="color: #00BB00;">“exAM_Microglia_Score”</span>.</span></span></code></pre>
<p><code>exAM_Scoring</code> also supports use of ensembl IDs by
specifying optional parameters.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">marsh_mouse_micro</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exAM_Scoring.html">exAM_Scoring</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">marsh_mouse_micro</span>, species <span class="op">=</span> <span class="st">"mouse"</span>, ensembl_ids <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="we-can-then-plot-the-score-using-any-of-the-normal-plotting-functions-">We can then plot the score using any of the normal plotting
functions.<a class="anchor" aria-label="anchor" href="#we-can-then-plot-the-score-using-any-of-the-normal-plotting-functions-"></a>
</h4>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DimPlot_scCustom.html">DimPlot_scCustom</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">marsh_mouse_micro</span>, colors_use <span class="op">=</span> <span class="va">marsh_mouse_micro</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">exp17_micro_colors</span>,</span>
<span>    label.size <span class="op">=</span> <span class="fl">5</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span>, repel <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/FeaturePlot_scCustom.html">FeaturePlot_scCustom</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">marsh_mouse_micro</span>, features <span class="op">=</span> <span class="st">"exAM_Union_Score"</span>, na_cutoff <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
&lt;img src=“/Users/samuelmarsh/Desktop/Bioinformatics
Software/scCustomize/docs/articles/Object_QC_Functions_files/figure-html/unnamed-chunk-29-1.png”
alt=“<em><strong>A.</strong> tSNE plot showing the clusters from Marsh
et al., 2022. <strong>B.</strong> FeaturePlot_scCustom view of the
module score showing expected localization in the ”Activated” exAM
cluster.</em>” width=“1248” /&gt;
<p class="caption">
<em><strong>A.</strong> tSNE plot showing the clusters from Marsh et
al., 2022. <strong>B.</strong> FeaturePlot_scCustom view of the module
score showing expected localization in the “Activated” exAM
cluster.</em>
</p>
</div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/VlnPlot_scCustom.html">VlnPlot_scCustom</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">marsh_mouse_micro</span>, features <span class="op">=</span> <span class="st">"exAM_Union_Score"</span>, colors_use <span class="op">=</span> <span class="va">marsh_mouse_micro</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">exp17_micro_colors</span><span class="op">)</span></span></code></pre></div>
<p><img src="Object_QC_Functions_files/figure-html/unnamed-chunk-30-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
