<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Plotting #2: QC Plots &amp; Analysis • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Plotting #2: QC Plots &amp; Analysis">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.2.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
    <li>
      <a href="../articles/cNMF_Functionality.html">cNMF Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="QC_Plots_files/kePrint-0.0.1/kePrint.js"></script><link href="QC_Plots_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Plotting #2: QC Plots &amp; Analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: December 11,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/articles/QC_Plots.Rmd" class="external-link"><code>vignettes/articles/QC_Plots.Rmd</code></a></small>
      <div class="hidden name"><code>QC_Plots.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="qc-metrics-plots">QC Metrics &amp; Plots<a class="anchor" aria-label="anchor" href="#qc-metrics-plots"></a>
</h2>
<p>One of the first steps in all scRNA-seq analyses is performing a
number of QC checks and plots so that data can be appropriately
filtered. scCustomize contains a number of functions that can be used to
quickly and easily generate some of the most relevant QC plots.</p>
<p>For details on functions for adding QC metrics and details about them
please see <a href="https://samuel-marsh.github.io/scCustomize/articles/Object_QC_Functions.html">Object
QC Vignette</a>.</p>
<p>For this tutorial, I will be utilizing HCA bone marrow cell data from
the SeuratData package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/qsbase/qs" class="external-link">qs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load Example Dataset</span></span>
<span><span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu">hcabm40k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/hcabm40k.SeuratData/man/hcabm40k.html" class="external-link">hcabm40k</a></span></span>
<span></span>
<span><span class="co"># Add pseudo group variable just for this vignette</span></span>
<span><span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM1"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span></span>
<span>    <span class="st">"MantonBM2"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM3"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Group 1"</span></span>
<span></span>
<span><span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM5"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span></span>
<span>    <span class="st">"MantonBM6"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM7"</span> <span class="op">|</span> <span class="va">hca_bm</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"MantonBM8"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Group 2"</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hca_bm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Cell_QC_Metrics.html">Add_Cell_QC_Metrics</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">hca_bm</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="plotting-qc-metrics">Plotting QC Metrics<a class="anchor" aria-label="anchor" href="#plotting-qc-metrics"></a>
</h3>
<p>scCustomize has a number of quick QC plotting options for ease of
use.<br><em>NOTE: Most scCustomize plotting functions contain <code>...</code>
parameter to allow user to supply any of the parameters for the original
Seurat function that is being used under the hood.</em></p>
</div>
<div class="section level3">
<h3 id="vlnplot-based-qc-plots">
<code>VlnPlot</code>-Based QC Plots<a class="anchor" aria-label="anchor" href="#vlnplot-based-qc-plots"></a>
</h3>
<p>scCustomize contains a number of shortcut/wrapper functions for QC
plotting, which are wrappers around
<code><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">Seurat::VlnPlot()</a></code>/<code>VlnPlot_scCustom</code>.</p>
<ul>
<li>
<code><a href="../reference/QC_Plots_Genes.html">QC_Plots_Genes()</a></code> Plots genes/features per
cell/nucleus.</li>
<li>
<code><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs()</a></code> Plots UMIs per cell/nucleus.<br>
</li>
<li>
<code><a href="../reference/QC_Plots_Mito.html">QC_Plots_Mito()</a></code> Plots mito% (named “percent_mito”) per
cell/nucleus.</li>
<li>
<code><a href="../reference/QC_Plots_Complexity.html">QC_Plots_Complexity()</a></code> Plots cell complexity metric
(log10GenesPerUMI) per cell/nucleus.</li>
<li>
<code><a href="../reference/QC_Plots_Feature.html">QC_Plots_Feature()</a></code> Plots “feature” per cell/nucleus.
Using parameter <code>feature</code> to allow plotting of any applicable
named feature in object@meta.data slot.</li>
<li>
<code><a href="../reference/QC_Plots_Combined_Vln.html">QC_Plots_Combined_Vln()</a></code> Returns patchwork plot of
<code><a href="../reference/QC_Plots_Genes.html">QC_Plots_Genes()</a></code>, <code><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs()</a></code>, &amp;
<code><a href="../reference/QC_Plots_Mito.html">QC_Plots_Mito()</a></code>.</li>
</ul>
<p>scCustomize functions have the added benefit of:</p>
<ul>
<li>Feature to plot set by default (except for
<code>QC_Plots_Feature</code>).</li>
<li>Added high/low cutoff parameters to allow for easy visualization of
potential cutoff thresholds.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All functions contain</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_Genes.html">QC_Plots_Genes</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">600</span>, high_cutoff <span class="op">=</span> <span class="fl">5500</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_Mito.html">QC_Plots_Mito</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, high_cutoff <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_Complexity.html">QC_Plots_Complexity</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, high_cutoff <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span>, ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-5-1.png" width="1248"></p>
<div class="section level4">
<h4 id="additional-parameters">Additional parameters<a class="anchor" aria-label="anchor" href="#additional-parameters"></a>
</h4>
<p>In addition to being able to supply Seurat parameters with
<code>...</code> these plots like many others in scCustomize contain
other additional parameters to customize plot output without need for
post-plot ggplot2 modifications</p>
<ul>
<li>
<code>plot_title</code>: Change plot title</li>
<li>
<code>x_axis_label</code>/<code>y_axis_label</code>: Change axis
labels.</li>
<li>
<code>x_lab_rotate</code>: Should x-axis label be rotated 45
degrees?</li>
<li>
<code>y_axis_log</code>: Should y-axis in linear or log10
scale.<br>
</li>
<li>
<code>plot_median</code> &amp; <code>median_size</code>: Plot a line
at the median of each x-axis identity.<br>
</li>
<li>
<code>plot_boxplot</code>: Plot boxplot on top of the violin.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    y_axis_log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-6-1.png" alt="*Setting `y_axis_log` can be very helpful for initial plots where outliers skew the visualization of the majority of the data without excluding data by setting y-axis limit.*" width="1248"><p class="caption">
<em>Setting <code>y_axis_log</code> can be very helpful for initial
plots where outliers skew the visualization of the majority of the data
without excluding data by setting y-axis limit.</em>
</p>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    plot_median <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    y_axis_log <span class="op">=</span> <span class="cn">TRUE</span>, plot_median <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-7-1.png" alt="*Plotting median values by setting the `plot_median = TRUE` parameter.*" width="1248"><p class="caption">
<em>Plotting median values by setting the
<code>plot_median = TRUE</code> parameter.</em>
</p>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    plot_boxplot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QC_Plots_UMIs.html">QC_Plots_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff <span class="op">=</span> <span class="fl">1200</span>, high_cutoff <span class="op">=</span> <span class="fl">45000</span>, pt.size <span class="op">=</span> <span class="fl">0</span>,</span>
<span>    y_axis_log <span class="op">=</span> <span class="cn">TRUE</span>, plot_boxplot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-8-1.png" alt="*Plotting boxplot by setting the `plot_boxplot = TRUE` parameter.*" width="1248"><p class="caption">
<em>Plotting boxplot by setting the <code>plot_boxplot = TRUE</code>
parameter.</em>
</p>
</div>
</div>
<div class="section level4">
<h4 id="combined-plotting-function">Combined Plotting Function<a class="anchor" aria-label="anchor" href="#combined-plotting-function"></a>
</h4>
<p>As a shortcut you can return single patchwork plot of the 3 main QC
Plots (Genes, UMIs, %Mito) by using single function,
<code><a href="../reference/QC_Plots_Combined_Vln.html">QC_Plots_Combined_Vln()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/QC_Plots_Combined_Vln.html">QC_Plots_Combined_Vln</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, feature_cutoffs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">600</span>, <span class="fl">5500</span><span class="op">)</span>, UMI_cutoffs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1200</span>,</span>
<span>    <span class="fl">45000</span><span class="op">)</span>, mito_cutoffs <span class="op">=</span> <span class="fl">20</span>, pt.size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-9-1.png" width="1248"></p>
</div>
</div>
<div class="section level3">
<h3 id="featurescatter-based-qc-plots">
<code>FeatureScatter</code>-Based QC Plots<a class="anchor" aria-label="anchor" href="#featurescatter-based-qc-plots"></a>
</h3>
<p>scCustomize contains 3 functions which wrap
<code><a href="https://satijalab.org/seurat/reference/FeatureScatter.html" class="external-link">Seurat::FeatureScatter()</a></code> with added visualization of
potential cutoff thresholds and some additional functionality:</p>
<ul>
<li>
<code><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> Plots genes vs UMIs per
cell/nucleus</li>
<li>
<code><a href="../reference/QC_Plot_GenevsFeature.html">QC_Plot_GenevsFeature()</a></code> Plots Genes vs. “feature” per
cell/nucleus. Using parameter <code>feature1</code> to allow plotting of
any applicable named feature in object@meta.data slot.<br>
</li>
<li>
<code><a href="../reference/QC_Plot_UMIvsFeature.html">QC_Plot_UMIvsFeature()</a></code> Plots UMIs vs. “feature” per
cell/nucleus. Using parameter <code>feature1</code> to allow plotting of
any applicable named feature in object@meta.data slot.</li>
</ul>
<div class="section level4">
<h4 id="newmodified-functionality">New/Modified functionality<a class="anchor" aria-label="anchor" href="#newmodified-functionality"></a>
</h4>
<ul>
<li>Better default color palettes</li>
<li>
<code>shuffle = TRUE</code> by default to prevent hiding of
datasets</li>
<li>Ability to set &amp; visualize potential cutoff thresholds (similar
to VlnPlot based QC Plots above)</li>
<li>Report potential post filtering correlation in addition to whole
dataset correlation when using <code>QC_Plot_UMIvsGene</code> (based on
values provided to high and low cutoff parameters)</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># All functions contain</span></span>
<span><span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">600</span>, high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, low_cutoff_UMI <span class="op">=</span> <span class="fl">500</span>,</span>
<span>    high_cutoff_UMI <span class="op">=</span> <span class="fl">50000</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/QC_Plot_GenevsFeature.html">QC_Plot_GenevsFeature</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, feature1 <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">600</span>,</span>
<span>    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_feature <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-11-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="color-data-by-continuous-meta-data-variable">Color data by continuous meta data variable<a class="anchor" aria-label="anchor" href="#color-data-by-continuous-meta-data-variable"></a>
</h4>
<p><code>QC_Plot_UMIvsGene</code> contains the ability to color points
by continuous meta data variables.<br>
This can be used to plot % of mito reads in addition to UMI vs. Gene
comparisons</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, meta_gradient_name <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">600</span>,</span>
<span>    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_UMI <span class="op">=</span> <span class="fl">45000</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, meta_gradient_name <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">600</span>,</span>
<span>    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_UMI <span class="op">=</span> <span class="fl">45000</span>, meta_gradient_low_cutoff <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-13-1.png" alt="*`QC_Plot_UMIvsGene()` when using `meta_gradient_name` outputs plot colored by meta data variable (left) to view only points above potential cutoff `meta_gradient_low_cutoff` can be specified to alter the plotting (right).*" width="1248"><p class="caption">
<em><code><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> when using
<code>meta_gradient_name</code> outputs plot colored by meta data
variable (left) to view only points above potential cutoff
<code>meta_gradient_low_cutoff</code> can be specified to alter the
plotting (right).</em>
</p>
</div>
</div>
<div class="section level4">
<h4 id="combination-plots">Combination Plots<a class="anchor" aria-label="anchor" href="#combination-plots"></a>
</h4>
<p>If you are interested in viewing <code>QC_Plot_UMIvsGene</code> both
by discrete grouping variable and by continuous variable without writing
function twice you can use <code>combination = TRUE</code> and plot
output will contain both plots.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, meta_gradient_name <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff_gene <span class="op">=</span> <span class="fl">600</span>,</span>
<span>    high_cutoff_gene <span class="op">=</span> <span class="fl">5500</span>, high_cutoff_UMI <span class="op">=</span> <span class="fl">45000</span>, meta_gradient_low_cutoff <span class="op">=</span> <span class="fl">20</span>, combination <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-14-1.png" alt="*`QC_Plot_UMIvsGene()` when using `combination = TRUE` will output both the Gene x UMI by active identity and with meta data gradient coloring.*" width="1248"><p class="caption">
<em><code><a href="../reference/QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> when using
<code>combination = TRUE</code> will output both the Gene x UMI by
active identity and with meta data gradient coloring.</em>
</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="histogram-based-qc-plots">Histogram-Based QC Plots<a class="anchor" aria-label="anchor" href="#histogram-based-qc-plots"></a>
</h3>
<p>Finally, scCustomize contains a function to plot QC metrics as
histogram instead of violin for visualization of the distribution of
feature.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/QC_Histogram.html">QC_Histogram</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, features <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/QC_Histogram.html">QC_Histogram</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, features <span class="op">=</span> <span class="st">"percent_mito"</span>, low_cutoff <span class="op">=</span> <span class="fl">15</span>, split.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="QC_Plots_files/figure-html/unnamed-chunk-16-1.png" alt="*`QC_Histogram` also supports splitting plots by meta.data variables*" width="1248"><p class="caption">
<em><code>QC_Histogram</code> also supports splitting plots by meta.data
variables</em>
</p>
</div>
</div>
<div class="section level3">
<h3 id="analyze-median-qc-values-per-samplelibrary">Analyze Median QC Values per Sample/Library<a class="anchor" aria-label="anchor" href="#analyze-median-qc-values-per-samplelibrary"></a>
</h3>
<p>scCustomize also contains a few helpful functions for returning and
plotting the median values for these metrics on per sample/library
basis.</p>
<div class="section level4">
<h4 id="calculate-median-values-return-data-frame">Calculate Median Values &amp; Return data.frame<a class="anchor" aria-label="anchor" href="#calculate-median-values-return-data-frame"></a>
</h4>
<p>scCustomize contains function <code>Median_Stats</code> to quickly
calculate the medians for basic QC stats (Genes/, UMIs/, %Mito/Cell,
etc) and return a data.frame.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
Median_nCount_RNA
</th>
<th style="text-align:right;">
Median_nFeature_RNA
</th>
<th style="text-align:right;">
Median_percent_mito
</th>
<th style="text-align:right;">
Median_percent_ribo
</th>
<th style="text-align:right;">
Median_percent_mito_ribo
</th>
<th style="text-align:right;">
Median_log10GenesPerUMI
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
MantonBM1
</td>
<td style="text-align:right;">
2634.5
</td>
<td style="text-align:right;">
807
</td>
<td style="text-align:right;">
3.928074
</td>
<td style="text-align:right;">
42.04921
</td>
<td style="text-align:right;">
46.43790
</td>
<td style="text-align:right;">
0.8506316
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM2
</td>
<td style="text-align:right;">
2462.5
</td>
<td style="text-align:right;">
783
</td>
<td style="text-align:right;">
3.556994
</td>
<td style="text-align:right;">
42.96184
</td>
<td style="text-align:right;">
46.87269
</td>
<td style="text-align:right;">
0.8505270
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM3
</td>
<td style="text-align:right;">
2617.0
</td>
<td style="text-align:right;">
769
</td>
<td style="text-align:right;">
3.301410
</td>
<td style="text-align:right;">
40.65558
</td>
<td style="text-align:right;">
44.10722
</td>
<td style="text-align:right;">
0.8483425
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM4
</td>
<td style="text-align:right;">
2264.0
</td>
<td style="text-align:right;">
780
</td>
<td style="text-align:right;">
3.384891
</td>
<td style="text-align:right;">
33.73245
</td>
<td style="text-align:right;">
37.96850
</td>
<td style="text-align:right;">
0.8645139
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM5
</td>
<td style="text-align:right;">
2677.0
</td>
<td style="text-align:right;">
741
</td>
<td style="text-align:right;">
3.265677
</td>
<td style="text-align:right;">
41.55203
</td>
<td style="text-align:right;">
45.33713
</td>
<td style="text-align:right;">
0.8360196
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM6
</td>
<td style="text-align:right;">
2677.0
</td>
<td style="text-align:right;">
835
</td>
<td style="text-align:right;">
3.202053
</td>
<td style="text-align:right;">
43.07874
</td>
<td style="text-align:right;">
46.70119
</td>
<td style="text-align:right;">
0.8522162
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM7
</td>
<td style="text-align:right;">
2429.0
</td>
<td style="text-align:right;">
732
</td>
<td style="text-align:right;">
3.216184
</td>
<td style="text-align:right;">
44.68831
</td>
<td style="text-align:right;">
48.20765
</td>
<td style="text-align:right;">
0.8477172
</td>
</tr>
<tr>
<td style="text-align:left;">
MantonBM8
</td>
<td style="text-align:right;">
2453.0
</td>
<td style="text-align:right;">
701
</td>
<td style="text-align:right;">
3.697331
</td>
<td style="text-align:right;">
44.97911
</td>
<td style="text-align:right;">
48.96420
</td>
<td style="text-align:right;">
0.8406128
</td>
</tr>
<tr>
<td style="text-align:left;">
Totals (All Cells)
</td>
<td style="text-align:right;">
2524.0
</td>
<td style="text-align:right;">
769
</td>
<td style="text-align:right;">
3.447040
</td>
<td style="text-align:right;">
41.53183
</td>
<td style="text-align:right;">
45.40989
</td>
<td style="text-align:right;">
0.8494067
</td>
</tr>
</tbody>
</table>
<p>The <code>Median_Stats</code> function has some column names stored
by default but will also calculate medians for additional meta.data
columns using the optional <code>median_var</code> parameter</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span>, median_var <span class="op">=</span> <span class="st">"meta_data_column_name"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plotting-median-values">Plotting Median Values<a class="anchor" aria-label="anchor" href="#plotting-median-values"></a>
</h4>
<p>scCustomize also contains a few functions to plot some of these
median value calculations, which can be used on their own without need
to return data.frame first.</p>
<ul>
<li>
<code><a href="../reference/Plot_Median_Genes.html">Plot_Median_Genes()</a></code><br>
</li>
<li>
<code><a href="../reference/Plot_Median_UMIs.html">Plot_Median_UMIs()</a></code><br>
</li>
<li>
<code><a href="../reference/Plot_Median_Mito.html">Plot_Median_Mito()</a></code><br>
</li>
<li>
<code><a href="../reference/Plot_Median_Other.html">Plot_Median_Other()</a></code>
<ul>
<li><em>Used to plot any other numeric variable present in object
meta.data slot.</em></li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Plot_Median_Genes.html">Plot_Median_Genes</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Plot_Median_UMIs.html">Plot_Median_UMIs</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Plot_Median_Mito.html">Plot_Median_Mito</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/Plot_Median_Other.html">Plot_Median_Other</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">hca_bm</span>, median_var <span class="op">=</span> <span class="st">"percent_ribo"</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-21-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="plot-number-of-cellsnuclei-per-sample">Plot Number of Cells/Nuclei per Sample<a class="anchor" aria-label="anchor" href="#plot-number-of-cellsnuclei-per-sample"></a>
</h4>
<p>scCustomize also contains plotting function to plot the number of
cells or nuclei per sample.</p>
<p>Since the HCA Bone Marrow dataset has exactly the same number of
cells per sample we will use the microglia object from the Analysis
Plots vignette.</p>
<p><img src="QC_Plots_files/figure-html/unnamed-chunk-23-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
