<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Plotting #3: Sequencing QC Plots/Analysis • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Plotting #3: Sequencing QC Plots/Analysis">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
    <li>
      <a href="../articles/cNMF_Functionality.html">cNMF Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Sequencing_QC_Plots_files/kePrint-0.0.1/kePrint.js"></script><link href="Sequencing_QC_Plots_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Plotting #3: Sequencing QC Plots/Analysis</h1>
            
            <h4 data-toc-skip class="date">Compiled: September 10,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/articles/Sequencing_QC_Plots.Rmd" class="external-link"><code>vignettes/articles/Sequencing_QC_Plots.Rmd</code></a></small>
      <div class="hidden name"><code>Sequencing_QC_Plots.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="plotting-sequencingalignment-metrics">Plotting Sequencing/Alignment Metrics<a class="anchor" aria-label="anchor" href="#plotting-sequencingalignment-metrics"></a>
</h2>
<p>In addition to other QC metrics previously discussed it can also be
helpful to plot the metrics generated during alignment of sequencing
data to see if there are differences that might arise as result of
sequencing differences as opposed to biological differences.</p>
<p>For this tutorial, I will be utilizing de-identified example data
from my own experience but I will walk through process for using your
own data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/qsbase/qs" class="external-link">qs</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="loading-data">Loading Data<a class="anchor" aria-label="anchor" href="#loading-data"></a>
</h3>
<p>Before we can plot these metrics we first need to import raw
data.</p>
<div class="section level4">
<h4 id="accepted-data-types">Accepted Data Types<a class="anchor" aria-label="anchor" href="#accepted-data-types"></a>
</h4>
<p>These functions currently only work with data from 10X Genomics
processed via Cell Ranger. If there is interest in adapting to other
platforms I’d be happy to entertain PRs.</p>
</div>
<div class="section level4">
<h4 id="data-location">Data Location<a class="anchor" aria-label="anchor" href="#data-location"></a>
</h4>
<p>Inside the output folder from Cell Ranger <code>count</code> is a
file called “metrics_summary” which contains the metrics normally
displayed in the html summary. See abbreviated example below.</p>
<pre><code>Parent_Directory
├── sample_01
│   └── outs
│       └── filtered_feature_bc_matrix/
│       └── raw_feature_bc_matrix/
│       └── web_summary.html
│       └── metrics_summary.csv
└── sample_02
│   └── outs
│       └── filtered_feature_bc_matrix/
│       └── raw_feature_bc_matrix/
│       └── web_summary.html
│       └── metrics_summary.csv
└── sample_03</code></pre>
<p><em>NOTE: If the path to these files on your system is different that
the default 10X path be sure to specify the <code>secondary_path</code>
param and set <code>default_10X = FALSE</code> when importing
data.</em></p>
</div>
<div class="section level4">
<h4 id="support-for-cell-ranger-multi-pipeline">Support for Cell Ranger <code>multi</code> pipeline<a class="anchor" aria-label="anchor" href="#support-for-cell-ranger-multi-pipeline"></a>
</h4>
<p><code>Read_Metrics_10X</code> includes support for data processed
using Cell Ranger <code>multi</code> pipeline by setting the parameter
<code>cellranger_multi = TRUE</code>. The function will automatically
alter it’s default secondary path value and will return both gene
expression and TCR metrics as separate data.frames.</p>
</div>
<div class="section level4">
<h4 id="read-data">Read Data<a class="anchor" aria-label="anchor" href="#read-data"></a>
</h4>
<p>We can read all metrics files within a given parent directory using
the <code>Read_Metrics_10X</code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">raw_metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Read_Metrics_10X.html">Read_Metrics_10X</a></span><span class="op">(</span>base_path <span class="op">=</span> <span class="st">"path/to/parent/directory/"</span>, default_10X <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">raw_metrics</span><span class="op">)</span></span></code></pre></div>
<p>The result of this function will be a data.frame with one row per
library.</p>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
sample_id
</th>
<th style="text-align:right;">
Estimated_Number_of_Cells
</th>
<th style="text-align:right;">
Mean_Reads_per_Cell
</th>
<th style="text-align:right;">
Median_Genes_per_Cell
</th>
<th style="text-align:right;">
Number_of_Reads
</th>
<th style="text-align:left;">
Valid_Barcodes
</th>
<th style="text-align:left;">
Sequencing_Saturation
</th>
<th style="text-align:left;">
Q30_Bases_in_Barcode
</th>
<th style="text-align:left;">
Q30_Bases_in_RNA_Read
</th>
<th style="text-align:left;">
Q30_Bases_in_UMI
</th>
<th style="text-align:left;">
Reads_Mapped_to_Genome
</th>
<th style="text-align:left;">
Reads_Mapped_Confidently_to_Genome
</th>
<th style="text-align:left;">
Reads_Mapped_Confidently_to_Intergenic_Regions
</th>
<th style="text-align:left;">
Reads_Mapped_Confidently_to_Intronic_Regions
</th>
<th style="text-align:left;">
Reads_Mapped_Confidently_to_Exonic_Regions
</th>
<th style="text-align:left;">
Reads_Mapped_Confidently_to_Transcriptome
</th>
<th style="text-align:left;">
Reads_Mapped_Antisense_to_Gene
</th>
<th style="text-align:left;">
Fraction_Reads_in_Cells
</th>
<th style="text-align:right;">
Total_Genes_Detected
</th>
<th style="text-align:right;">
Median_UMI_Counts_per_Cell
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
sample_1
</td>
<td style="text-align:right;">
2526
</td>
<td style="text-align:right;">
193933
</td>
<td style="text-align:right;">
724
</td>
<td style="text-align:right;">
489875507
</td>
<td style="text-align:left;">
96.3%
</td>
<td style="text-align:left;">
89.8%
</td>
<td style="text-align:left;">
96.0%
</td>
<td style="text-align:left;">
94.2%
</td>
<td style="text-align:left;">
95.7%
</td>
<td style="text-align:left;">
95.9%
</td>
<td style="text-align:left;">
86.5%
</td>
<td style="text-align:left;">
5.2%
</td>
<td style="text-align:left;">
21.7%
</td>
<td style="text-align:left;">
59.5%
</td>
<td style="text-align:left;">
56.1%
</td>
<td style="text-align:left;">
2.3%
</td>
<td style="text-align:left;">
60.1%
</td>
<td style="text-align:right;">
21105
</td>
<td style="text-align:right;">
1840
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_2
</td>
<td style="text-align:right;">
2642
</td>
<td style="text-align:right;">
179146
</td>
<td style="text-align:right;">
986
</td>
<td style="text-align:right;">
473302676
</td>
<td style="text-align:left;">
97.7%
</td>
<td style="text-align:left;">
88.3%
</td>
<td style="text-align:left;">
95.9%
</td>
<td style="text-align:left;">
94.3%
</td>
<td style="text-align:left;">
95.6%
</td>
<td style="text-align:left;">
95.8%
</td>
<td style="text-align:left;">
88.7%
</td>
<td style="text-align:left;">
4.6%
</td>
<td style="text-align:left;">
18.4%
</td>
<td style="text-align:left;">
65.7%
</td>
<td style="text-align:left;">
63.0%
</td>
<td style="text-align:left;">
1.7%
</td>
<td style="text-align:left;">
56.7%
</td>
<td style="text-align:right;">
21321
</td>
<td style="text-align:right;">
3394
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_3
</td>
<td style="text-align:right;">
7337
</td>
<td style="text-align:right;">
79420
</td>
<td style="text-align:right;">
2454
</td>
<td style="text-align:right;">
582707755
</td>
<td style="text-align:left;">
93.5%
</td>
<td style="text-align:left;">
72.3%
</td>
<td style="text-align:left;">
96.5%
</td>
<td style="text-align:left;">
94.8%
</td>
<td style="text-align:left;">
96.2%
</td>
<td style="text-align:left;">
95.5%
</td>
<td style="text-align:left;">
90.8%
</td>
<td style="text-align:left;">
7.1%
</td>
<td style="text-align:left;">
20.1%
</td>
<td style="text-align:left;">
63.7%
</td>
<td style="text-align:left;">
58.3%
</td>
<td style="text-align:left;">
4.3%
</td>
<td style="text-align:left;">
72.5%
</td>
<td style="text-align:right;">
23556
</td>
<td style="text-align:right;">
6685
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_4
</td>
<td style="text-align:right;">
2352
</td>
<td style="text-align:right;">
165755
</td>
<td style="text-align:right;">
722
</td>
<td style="text-align:right;">
389854975
</td>
<td style="text-align:left;">
97.0%
</td>
<td style="text-align:left;">
86.9%
</td>
<td style="text-align:left;">
96.0%
</td>
<td style="text-align:left;">
94.1%
</td>
<td style="text-align:left;">
95.6%
</td>
<td style="text-align:left;">
95.0%
</td>
<td style="text-align:left;">
84.0%
</td>
<td style="text-align:left;">
4.8%
</td>
<td style="text-align:left;">
20.3%
</td>
<td style="text-align:left;">
58.9%
</td>
<td style="text-align:left;">
55.7%
</td>
<td style="text-align:left;">
2.2%
</td>
<td style="text-align:left;">
66.4%
</td>
<td style="text-align:right;">
21141
</td>
<td style="text-align:right;">
2360
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_5
</td>
<td style="text-align:right;">
3697
</td>
<td style="text-align:right;">
152621
</td>
<td style="text-align:right;">
794
</td>
<td style="text-align:right;">
564238318
</td>
<td style="text-align:left;">
97.7%
</td>
<td style="text-align:left;">
89.2%
</td>
<td style="text-align:left;">
96.0%
</td>
<td style="text-align:left;">
94.3%
</td>
<td style="text-align:left;">
95.7%
</td>
<td style="text-align:left;">
96.3%
</td>
<td style="text-align:left;">
85.7%
</td>
<td style="text-align:left;">
3.8%
</td>
<td style="text-align:left;">
20.9%
</td>
<td style="text-align:left;">
61.0%
</td>
<td style="text-align:left;">
58.1%
</td>
<td style="text-align:left;">
1.9%
</td>
<td style="text-align:left;">
66.3%
</td>
<td style="text-align:right;">
21764
</td>
<td style="text-align:right;">
2224
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_6
</td>
<td style="text-align:right;">
5043
</td>
<td style="text-align:right;">
101180
</td>
<td style="text-align:right;">
1126
</td>
<td style="text-align:right;">
510250215
</td>
<td style="text-align:left;">
94.5%
</td>
<td style="text-align:left;">
74.6%
</td>
<td style="text-align:left;">
96.4%
</td>
<td style="text-align:left;">
94.8%
</td>
<td style="text-align:left;">
96.0%
</td>
<td style="text-align:left;">
96.1%
</td>
<td style="text-align:left;">
81.9%
</td>
<td style="text-align:left;">
5.1%
</td>
<td style="text-align:left;">
15.8%
</td>
<td style="text-align:left;">
61.0%
</td>
<td style="text-align:left;">
56.5%
</td>
<td style="text-align:left;">
3.5%
</td>
<td style="text-align:left;">
78.5%
</td>
<td style="text-align:right;">
22565
</td>
<td style="text-align:right;">
8792
</td>
</tr>
</tbody>
</table>
<div class="section level5">
<h5 id="optional-parameters">Optional parameters<a class="anchor" aria-label="anchor" href="#optional-parameters"></a>
</h5>
<p>Similarly to the scCustomize functions for reading raw count matrices
(See <a href="https://samuel-marsh.github.io/scCustomize/articles/Read_and_Write_Functions.html">Read
&amp; Write Data Functions Vignette</a>) there are a few optional
parameters for <code>Read_Metrics_10X</code>.</p>
<ul>
<li>
<code>secondary_path</code>/<code>default_10X</code> tells the
function how to navigate from the parent directory
(<code>base_path</code>) to the “metrics_summary.csv” files.
<ul>
<li>If the directory specified is directly from Cell Ranger output or
has an identical structure (see above; parent/library_name/outs/files)
then user can specify <code>default_10X = TRUE</code> and function will
find required files.</li>
<li>If structure is different then user must supply
<code>secondary_path</code>.<br>
</li>
<li>If all files are in one directory simply supply
<code>secondary_path = ""</code>.<br>
</li>
</ul>
</li>
<li>
<code>lib_list</code> if only a subset of libraries are desired then
user can supply the names of those libraries (sub-directories) to the
<code>lib_list</code> parameter to limit which files are read in.
Default is NULL and will read all samples from parent directory.</li>
<li>
<code>lib_names</code> How to name files in output data.frame.
Default will use sub-directory names to label the samples. Alternatively
can provide a vector a sample/library names to use instead.</li>
</ul>
</div>
</div>
<div class="section level4">
<h4 id="add-meta-data">Add Meta Data<a class="anchor" aria-label="anchor" href="#add-meta-data"></a>
</h4>
<p>In order to make the resulting plots more meaningful we need to added
corresponding sample level meta data.<br><em>NOTE: Make sure that one column in your meta data data.frame matches
the “sample_id” column to metrics data.frame so that they are properly
joined.</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta_metrics</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
sample_id
</th>
<th style="text-align:left;">
Seq_Batch
</th>
<th style="text-align:right;">
Exp_Batch
</th>
<th style="text-align:left;">
Full_Batch
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
sample_1
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
1_A
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_2
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
2_A
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_3
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
3_B
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_4
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:left;">
1_A
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_5
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:left;">
2_A
</td>
</tr>
<tr>
<td style="text-align:left;">
sample_6
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:left;">
3_B
</td>
</tr>
</tbody>
</table>
<p>Now we join the two data.frames before plotting.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metrics_final</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">right_join</a></span><span class="op">(</span><span class="va">raw_metrics</span>, <span class="va">meta_metrics</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(sample_id)`</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h3>
<p>In total there are 15 Sequencing QC plotting functions. See <a href="https://samuel-marsh.github.io/scCustomize/reference/index.html#section-sequencing-metrics-qc-plots">Reference
Manual</a> for full details.</p>
<p>However, to simplify plotting there are also two wrapper functions
which will return patchwork layouts of several of these functions in a
single plot.</p>
<div class="section level4">
<h4 id="basic-metrics-summary">Basic Metrics Summary<a class="anchor" aria-label="anchor" href="#basic-metrics-summary"></a>
</h4>
<p>A set of basic metrics can be plotted using
<code>Seq_QC_Plot_Basic_Combined</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Seq_QC_Plot_Basic_Combined.html">Seq_QC_Plot_Basic_Combined</a></span><span class="op">(</span>metrics_dataframe <span class="op">=</span> <span class="va">metrics_final</span>, plot_by <span class="op">=</span> <span class="st">"Full_Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sequencing_QC_Plots_files/figure-html/unnamed-chunk-7-1.png" width="1248"></p>
</div>
<div class="section level4">
<h4 id="alignment-metrics-summary">Alignment Metrics Summary<a class="anchor" aria-label="anchor" href="#alignment-metrics-summary"></a>
</h4>
<p>A set of alignment-based metrics can be plotted using
<code>Seq_QC_Plot_Alignment_Combined</code></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Seq_QC_Plot_Alignment_Combined.html">Seq_QC_Plot_Alignment_Combined</a></span><span class="op">(</span>metrics_dataframe <span class="op">=</span> <span class="va">metrics_final</span>, plot_by <span class="op">=</span> <span class="st">"Full_Batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sequencing_QC_Plots_files/figure-html/unnamed-chunk-8-1.png" width="1248"></p>
</div>
<div class="section level4">
<h4 id="basic-statistics">Basic Statistics<a class="anchor" aria-label="anchor" href="#basic-statistics"></a>
</h4>
<p>All of the <code>Seq_QC_Plot_...</code> functions also contain
parameter to calculate significance between all groups using
<code><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html" class="external-link">ggpubr::stat_compare_means()</a></code>.<br><strong><em>NOTE: See function help pages and ggpubr for more info to
make sure the right test is being used for your data.</em></strong></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Seq_QC_Plot_Basic_Combined.html">Seq_QC_Plot_Basic_Combined</a></span><span class="op">(</span>metrics_dataframe <span class="op">=</span> <span class="va">metrics_final</span>, plot_by <span class="op">=</span> <span class="st">"Full_Batch"</span>, significance <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="Sequencing_QC_Plots_files/figure-html/unnamed-chunk-9-1.png" width="1248"></p>
</div>
<div class="section level4">
<h4 id="barcode-rank-plots">Barcode Rank Plots<a class="anchor" aria-label="anchor" href="#barcode-rank-plots"></a>
</h4>
<p>scCustomize can also be used to create customized barcode rank plots
from raw feature barcode data using outputs from DropletUtils package.
scCustomize provides the function <code>Iterate_Barcode_Rank_Plot</code>
which reads data, processes using DropletUtils, and returns plots.</p>
<p>The plots are adaptation of the plots from DropletUtils package moved
to ggplot2 framework with some additional customization and the ability
to return rastered plots.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Iterate_Barcode_Rank_Plot.html">Iterate_Barcode_Rank_Plot</a></span><span class="op">(</span>dir_path_h5 <span class="op">=</span> <span class="st">"assets/Barcode_Rank_Example/"</span>, plateau <span class="op">=</span> <span class="fl">20000</span>, file_path <span class="op">=</span> <span class="st">"plots/"</span>,</span>
<span>    file_name <span class="op">=</span> <span class="st">"Barcode_Rank_Plots.pdf"</span><span class="op">)</span></span></code></pre></div>
<center>
<img src="../reference/figures/assets/Barcode_Rank_Plot_Example.jpg" width="75%"><em>Example of barcode rank plot output. (10K PBMCs from 10X Genomics
Public Data).</em>
</center>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
