<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Statistics Functions • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Statistics Functions">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Statistics_files/kePrint-0.0.1/kePrint.js"></script><link href="Statistics_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Statistics Functions</h1>
            
            <h4 data-toc-skip class="date">Compiled: August 26,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/articles/Statistics.Rmd" class="external-link"><code>vignettes/articles/Statistics.Rmd</code></a></small>
      <div class="hidden name"><code>Statistics.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="statistics-functions">Statistics Functions<a class="anchor" aria-label="anchor" href="#statistics-functions"></a>
</h2>
<p>scCustomize contains a couple simple helper functions to return
metrics of interest.</p>
<p>For this vignette I will be utilizing pbmc3k dataset from the
SeuratData package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/qsbase/qs" class="external-link">qs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load example dataset for tutorial</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">pbmc3k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/pbmc3k.SeuratData/man/pbmc3k.final.html" class="external-link">pbmc3k.final</a></span></span></code></pre></div>
<p>Now let’s add some extra meta data for use with tutorial</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add mito and ribo data</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Add_Mito_Ribo.html">Add_Mito_Ribo</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, species <span class="op">=</span> <span class="st">"human"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add random sample and group variables</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span>, <span class="st">"sample3"</span>, <span class="st">"sample4"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample1"</span> <span class="op">|</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Group 1"</span></span>
<span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">group</span><span class="op">[</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample2"</span> <span class="op">|</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample4"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Group 2"</span></span>
<span></span>
<span></span>
<span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">treatment</span><span class="op">[</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Control"</span></span>
<span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">treatment</span><span class="op">[</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample2"</span> <span class="op">|</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample4"</span> <span class="op">|</span></span>
<span>    <span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span> <span class="op">==</span> <span class="st">"sample3"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Treated"</span></span>
<span></span>
<span><span class="co"># Add dummy module score</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AddModuleScore.html" class="external-link">AddModuleScore</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3E"</span>, <span class="st">"CD4"</span>, <span class="st">"THY1"</span>, <span class="st">"TCRA"</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"module_score"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="cells-per-identity">Cells Per Identity<a class="anchor" aria-label="anchor" href="#cells-per-identity"></a>
</h3>
<p>It can be really helpful to know the number and percentage of cells
per identity/cluster during analysis.<br>
scCustomize provides the <code>Cluster_Stats_All_Samples</code> function
to make this easy.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Cluster_Stats_All_Samples.html">Cluster_Stats_All_Samples</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_stats</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
Cluster
</th>
<th style="text-align:right;">
Number
</th>
<th style="text-align:right;">
Freq
</th>
<th style="text-align:right;">
sample1
</th>
<th style="text-align:right;">
sample2
</th>
<th style="text-align:right;">
sample3
</th>
<th style="text-align:right;">
sample4
</th>
<th style="text-align:right;">
sample1_%
</th>
<th style="text-align:right;">
sample2_%
</th>
<th style="text-align:right;">
sample3_%
</th>
<th style="text-align:right;">
sample4_%
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:right;">
697
</td>
<td style="text-align:right;">
26.4215315
</td>
<td style="text-align:right;">
182
</td>
<td style="text-align:right;">
176
</td>
<td style="text-align:right;">
179
</td>
<td style="text-align:right;">
160
</td>
<td style="text-align:right;">
26.9629630
</td>
<td style="text-align:right;">
26.3079223
</td>
<td style="text-align:right;">
27.2036474
</td>
<td style="text-align:right;">
25.1572327
</td>
</tr>
<tr>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:right;">
483
</td>
<td style="text-align:right;">
18.3093252
</td>
<td style="text-align:right;">
141
</td>
<td style="text-align:right;">
127
</td>
<td style="text-align:right;">
112
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
20.8888889
</td>
<td style="text-align:right;">
18.9835575
</td>
<td style="text-align:right;">
17.0212766
</td>
<td style="text-align:right;">
16.1949686
</td>
</tr>
<tr>
<td style="text-align:left;">
CD14+ Mono
</td>
<td style="text-align:right;">
480
</td>
<td style="text-align:right;">
18.1956027
</td>
<td style="text-align:right;">
118
</td>
<td style="text-align:right;">
126
</td>
<td style="text-align:right;">
117
</td>
<td style="text-align:right;">
119
</td>
<td style="text-align:right;">
17.4814815
</td>
<td style="text-align:right;">
18.8340807
</td>
<td style="text-align:right;">
17.7811550
</td>
<td style="text-align:right;">
18.7106918
</td>
</tr>
<tr>
<td style="text-align:left;">
B
</td>
<td style="text-align:right;">
344
</td>
<td style="text-align:right;">
13.0401820
</td>
<td style="text-align:right;">
79
</td>
<td style="text-align:right;">
90
</td>
<td style="text-align:right;">
78
</td>
<td style="text-align:right;">
97
</td>
<td style="text-align:right;">
11.7037037
</td>
<td style="text-align:right;">
13.4529148
</td>
<td style="text-align:right;">
11.8541033
</td>
<td style="text-align:right;">
15.2515723
</td>
</tr>
<tr>
<td style="text-align:left;">
CD8 T
</td>
<td style="text-align:right;">
271
</td>
<td style="text-align:right;">
10.2729340
</td>
<td style="text-align:right;">
60
</td>
<td style="text-align:right;">
65
</td>
<td style="text-align:right;">
78
</td>
<td style="text-align:right;">
68
</td>
<td style="text-align:right;">
8.8888889
</td>
<td style="text-align:right;">
9.7159940
</td>
<td style="text-align:right;">
11.8541033
</td>
<td style="text-align:right;">
10.6918239
</td>
</tr>
<tr>
<td style="text-align:left;">
FCGR3A+ Mono
</td>
<td style="text-align:right;">
162
</td>
<td style="text-align:right;">
6.1410159
</td>
<td style="text-align:right;">
42
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
40
</td>
<td style="text-align:right;">
46
</td>
<td style="text-align:right;">
6.2222222
</td>
<td style="text-align:right;">
5.0822123
</td>
<td style="text-align:right;">
6.0790274
</td>
<td style="text-align:right;">
7.2327044
</td>
</tr>
<tr>
<td style="text-align:left;">
NK
</td>
<td style="text-align:right;">
155
</td>
<td style="text-align:right;">
5.8756634
</td>
<td style="text-align:right;">
43
</td>
<td style="text-align:right;">
37
</td>
<td style="text-align:right;">
41
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
6.3703704
</td>
<td style="text-align:right;">
5.5306428
</td>
<td style="text-align:right;">
6.2310030
</td>
<td style="text-align:right;">
5.3459119
</td>
</tr>
<tr>
<td style="text-align:left;">
DC
</td>
<td style="text-align:right;">
32
</td>
<td style="text-align:right;">
1.2130402
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
10
</td>
<td style="text-align:right;">
9
</td>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
1.0370370
</td>
<td style="text-align:right;">
1.4947683
</td>
<td style="text-align:right;">
1.3677812
</td>
<td style="text-align:right;">
0.9433962
</td>
</tr>
<tr>
<td style="text-align:left;">
Platelet
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
0.5307051
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
0.4444444
</td>
<td style="text-align:right;">
0.5979073
</td>
<td style="text-align:right;">
0.6079027
</td>
<td style="text-align:right;">
0.4716981
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:right;">
2638
</td>
<td style="text-align:right;">
100.0000000
</td>
<td style="text-align:right;">
675
</td>
<td style="text-align:right;">
669
</td>
<td style="text-align:right;">
658
</td>
<td style="text-align:right;">
636
</td>
<td style="text-align:right;">
100.0000000
</td>
<td style="text-align:right;">
100.0000000
</td>
<td style="text-align:right;">
100.0000000
</td>
<td style="text-align:right;">
100.0000000
</td>
</tr>
</tbody>
</table>
<p>By default <code>Cluster_Stats_All_Samples</code> uses “orig.ident”
as the group variable but user can specify any <code>@meta.data</code>
slot variable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Cluster_Stats_All_Samples.html">Cluster_Stats_All_Samples</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cluster_stats</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
Cluster
</th>
<th style="text-align:right;">
Number
</th>
<th style="text-align:right;">
Freq
</th>
<th style="text-align:right;">
Group 1
</th>
<th style="text-align:right;">
Group 2
</th>
<th style="text-align:right;">
Group 1_%
</th>
<th style="text-align:right;">
Group 2_%
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Naive CD4 T
</td>
<td style="text-align:right;">
697
</td>
<td style="text-align:right;">
26.4215315
</td>
<td style="text-align:right;">
361
</td>
<td style="text-align:right;">
336
</td>
<td style="text-align:right;">
27.0817704
</td>
<td style="text-align:right;">
25.7471264
</td>
</tr>
<tr>
<td style="text-align:left;">
Memory CD4 T
</td>
<td style="text-align:right;">
483
</td>
<td style="text-align:right;">
18.3093252
</td>
<td style="text-align:right;">
253
</td>
<td style="text-align:right;">
230
</td>
<td style="text-align:right;">
18.9797449
</td>
<td style="text-align:right;">
17.6245211
</td>
</tr>
<tr>
<td style="text-align:left;">
CD14+ Mono
</td>
<td style="text-align:right;">
480
</td>
<td style="text-align:right;">
18.1956027
</td>
<td style="text-align:right;">
235
</td>
<td style="text-align:right;">
245
</td>
<td style="text-align:right;">
17.6294074
</td>
<td style="text-align:right;">
18.7739464
</td>
</tr>
<tr>
<td style="text-align:left;">
B
</td>
<td style="text-align:right;">
344
</td>
<td style="text-align:right;">
13.0401820
</td>
<td style="text-align:right;">
157
</td>
<td style="text-align:right;">
187
</td>
<td style="text-align:right;">
11.7779445
</td>
<td style="text-align:right;">
14.3295019
</td>
</tr>
<tr>
<td style="text-align:left;">
CD8 T
</td>
<td style="text-align:right;">
271
</td>
<td style="text-align:right;">
10.2729340
</td>
<td style="text-align:right;">
138
</td>
<td style="text-align:right;">
133
</td>
<td style="text-align:right;">
10.3525881
</td>
<td style="text-align:right;">
10.1915709
</td>
</tr>
<tr>
<td style="text-align:left;">
FCGR3A+ Mono
</td>
<td style="text-align:right;">
162
</td>
<td style="text-align:right;">
6.1410159
</td>
<td style="text-align:right;">
82
</td>
<td style="text-align:right;">
80
</td>
<td style="text-align:right;">
6.1515379
</td>
<td style="text-align:right;">
6.1302682
</td>
</tr>
<tr>
<td style="text-align:left;">
NK
</td>
<td style="text-align:right;">
155
</td>
<td style="text-align:right;">
5.8756634
</td>
<td style="text-align:right;">
84
</td>
<td style="text-align:right;">
71
</td>
<td style="text-align:right;">
6.3015754
</td>
<td style="text-align:right;">
5.4406130
</td>
</tr>
<tr>
<td style="text-align:left;">
DC
</td>
<td style="text-align:right;">
32
</td>
<td style="text-align:right;">
1.2130402
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
16
</td>
<td style="text-align:right;">
1.2003001
</td>
<td style="text-align:right;">
1.2260536
</td>
</tr>
<tr>
<td style="text-align:left;">
Platelet
</td>
<td style="text-align:right;">
14
</td>
<td style="text-align:right;">
0.5307051
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
7
</td>
<td style="text-align:right;">
0.5251313
</td>
<td style="text-align:right;">
0.5363985
</td>
</tr>
<tr>
<td style="text-align:left;">
Total
</td>
<td style="text-align:right;">
2638
</td>
<td style="text-align:right;">
100.0000000
</td>
<td style="text-align:right;">
1333
</td>
<td style="text-align:right;">
1305
</td>
<td style="text-align:right;">
100.0000000
</td>
<td style="text-align:right;">
100.0000000
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="plotting-clusteridentity-proportions">Plotting Cluster/Identity Proportions<a class="anchor" aria-label="anchor" href="#plotting-clusteridentity-proportions"></a>
</h3>
<p>Understanding the proportion of cells per cluster and how that may
change across a group can be important aspect of single cell analysis.
scCustomize contains function <code>Proportion_Plot</code> to provide
some visualization of cell proportions.</p>
<p>Plotting proportionality can be tricky especially when dealing with
clusters of very disparate sizes (especially small clusters) so plese be
aware to examine raw data as well (see
<code><a href="../reference/Cluster_Stats_All_Samples.html">Cluster_Stats_All_Samples()</a></code>).</p>
<p>Also note that when splitting plot by group there are now error bars
on these plots for interpretation of variance and other plotting methods
should be used to convery such information.</p>
<div class="section level4">
<h4 id="basic-proportion-plots">Basic Proportion Plots<a class="anchor" aria-label="anchor" href="#basic-proportion-plots"></a>
</h4>
<p><code>Proportion_Plot</code> can either return a bar graph or pie
graph depending on parameter value for</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Proportion_Plot.html">Proportion_Plot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Proportion_Plot.html">Proportion_Plot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, plot_type <span class="op">=</span> <span class="st">"pie"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statistics_files/figure-html/unnamed-chunk-8-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="plot-across-groups">Plot across groups<a class="anchor" aria-label="anchor" href="#plot-across-groups"></a>
</h4>
<p>Like other plots you can also use <code>Proportion_Plot</code> to
split plot across any categorical meta.data variable.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Proportion_Plot.html">Proportion_Plot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span>, split.by <span class="op">=</span> <span class="st">"treatment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statistics_files/figure-html/unnamed-chunk-9-1.png" width="1248" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Proportion_Plot.html">Proportion_Plot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, plot_type <span class="op">=</span> <span class="st">"pie"</span>, split.by <span class="op">=</span> <span class="st">"treatment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statistics_files/figure-html/unnamed-chunk-10-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="plot-raw-cell-numbers">Plot raw cell numbers<a class="anchor" aria-label="anchor" href="#plot-raw-cell-numbers"></a>
</h4>
<p>The default for <code>Proportion_Plot</code> is to plot the
proportion of cells and not the raw number of cells per identity because
that can be deceptive. However, in some cases it can be useful and so
users can change the <code>plot_scale</code> parameter to plot raw
numbers of cells instead of proportions.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Proportion_Plot.html">Proportion_Plot</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, plot_type <span class="op">=</span> <span class="st">"bar"</span>, split.by <span class="op">=</span> <span class="st">"treatment"</span>, plot_scale <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statistics_files/figure-html/unnamed-chunk-11-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="percent-of-cells-expressing-features">Percent of Cells Expressing Feature(s)<a class="anchor" aria-label="anchor" href="#percent-of-cells-expressing-features"></a>
</h3>
<p>It can also be informative to understand the percent of cells/nuclei
that express a given feature or set of features.<br>
scCustomize provides the <code>Percent_Expressing</code> function to
return these results.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percent_express</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Percent_Expressing.html">Percent_Expressing</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Naive.CD4.T
</th>
<th style="text-align:right;">
Memory.CD4.T
</th>
<th style="text-align:right;">
CD14..Mono
</th>
<th style="text-align:right;">
B
</th>
<th style="text-align:right;">
CD8.T
</th>
<th style="text-align:right;">
FCGR3A..Mono
</th>
<th style="text-align:right;">
NK
</th>
<th style="text-align:right;">
DC
</th>
<th style="text-align:right;">
Platelet
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
CD4
</td>
<td style="text-align:right;">
6.743185
</td>
<td style="text-align:right;">
15.942029
</td>
<td style="text-align:right;">
24.375
</td>
<td style="text-align:right;">
1.453488
</td>
<td style="text-align:right;">
2.583026
</td>
<td style="text-align:right;">
27.160494
</td>
<td style="text-align:right;">
0.6451613
</td>
<td style="text-align:right;">
34.375
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CD8A
</td>
<td style="text-align:right;">
13.342898
</td>
<td style="text-align:right;">
8.902691
</td>
<td style="text-align:right;">
1.875
</td>
<td style="text-align:right;">
2.616279
</td>
<td style="text-align:right;">
50.553505
</td>
<td style="text-align:right;">
3.703704
</td>
<td style="text-align:right;">
8.3870968
</td>
<td style="text-align:right;">
3.125
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
<div class="section level4">
<h4 id="change-grouping-variable">Change grouping variable<a class="anchor" aria-label="anchor" href="#change-grouping-variable"></a>
</h4>
<p>By default the function groups expression across
<code>@active.ident</code> (see above) but user can specify difference
variable if desired.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percent_express</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Percent_Expressing.html">Percent_Expressing</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8A"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
sample3
</th>
<th style="text-align:right;">
sample1
</th>
<th style="text-align:right;">
sample4
</th>
<th style="text-align:right;">
sample2
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
CD4
</td>
<td style="text-align:right;">
11.39818
</td>
<td style="text-align:right;">
11.55556
</td>
<td style="text-align:right;">
11.79245
</td>
<td style="text-align:right;">
12.10762
</td>
</tr>
<tr>
<td style="text-align:left;">
CD8A
</td>
<td style="text-align:right;">
12.91793
</td>
<td style="text-align:right;">
12.74074
</td>
<td style="text-align:right;">
11.00629
</td>
<td style="text-align:right;">
10.46338
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="split-within-groups">Split within groups<a class="anchor" aria-label="anchor" href="#split-within-groups"></a>
</h4>
<p>User can also supply a <code>split_by</code> variable to quantify
expression within group split by meta data variable</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percent_express</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Percent_Expressing.html">Percent_Expressing</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8A"</span><span class="op">)</span>, split_by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Naive.CD4.T
</th>
<th style="text-align:right;">
Memory.CD4.T
</th>
<th style="text-align:right;">
CD14..Mono
</th>
<th style="text-align:right;">
B
</th>
<th style="text-align:right;">
CD8.T
</th>
<th style="text-align:right;">
FCGR3A..Mono
</th>
<th style="text-align:right;">
NK
</th>
<th style="text-align:right;">
DC
</th>
<th style="text-align:right;">
Platelet
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
CD4
</td>
<td style="text-align:right;">
6.743185
</td>
<td style="text-align:right;">
15.942029
</td>
<td style="text-align:right;">
24.375
</td>
<td style="text-align:right;">
1.453488
</td>
<td style="text-align:right;">
2.583026
</td>
<td style="text-align:right;">
27.160494
</td>
<td style="text-align:right;">
0.6451613
</td>
<td style="text-align:right;">
34.375
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CD8A
</td>
<td style="text-align:right;">
13.342898
</td>
<td style="text-align:right;">
8.902691
</td>
<td style="text-align:right;">
1.875
</td>
<td style="text-align:right;">
2.616279
</td>
<td style="text-align:right;">
50.553505
</td>
<td style="text-align:right;">
3.703704
</td>
<td style="text-align:right;">
8.3870968
</td>
<td style="text-align:right;">
3.125
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="set-threshold-of-expression">Set threshold of expression<a class="anchor" aria-label="anchor" href="#set-threshold-of-expression"></a>
</h4>
<p>Users can also supply a threshold value to quantify percent of cells
expressing feature(s) above a certain threshold (be sure to note which
slot is being used to quantify expression when setting
thresholds).<br><em>NOTE: <code>Percent_Expressing</code> currently only supports single
threshold across all features</em></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percent_express</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Percent_Expressing.html">Percent_Expressing</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD4"</span>, <span class="st">"CD8A"</span><span class="op">)</span>, threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Naive.CD4.T
</th>
<th style="text-align:right;">
Memory.CD4.T
</th>
<th style="text-align:right;">
CD14..Mono
</th>
<th style="text-align:right;">
B
</th>
<th style="text-align:right;">
CD8.T
</th>
<th style="text-align:right;">
FCGR3A..Mono
</th>
<th style="text-align:right;">
NK
</th>
<th style="text-align:right;">
DC
</th>
<th style="text-align:right;">
Platelet
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
CD4
</td>
<td style="text-align:right;">
1.147776
</td>
<td style="text-align:right;">
0.621118
</td>
<td style="text-align:right;">
5.8333333
</td>
<td style="text-align:right;">
0.5813953
</td>
<td style="text-align:right;">
0.3690037
</td>
<td style="text-align:right;">
1.851852
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
3.125
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
CD8A
</td>
<td style="text-align:right;">
6.456241
</td>
<td style="text-align:right;">
2.070393
</td>
<td style="text-align:right;">
0.2083333
</td>
<td style="text-align:right;">
0.8720930
</td>
<td style="text-align:right;">
24.7232472
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
3.225807
</td>
<td style="text-align:right;">
0.000
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="calculate-summary-median-values">Calculate Summary Median Values<a class="anchor" aria-label="anchor" href="#calculate-summary-median-values"></a>
</h3>
<p>scCustomize contains function <code>Median_Stats</code> to quickly
calculate the medians for basic QC stats (Genes/, UMIs/, %
Mito/Cell).</p>
<div class="section level4">
<h4 id="basic-use">Basic Use<a class="anchor" aria-label="anchor" href="#basic-use"></a>
</h4>
<p>By default <code>Median_Stats</code> will calculate medians for the
following meta data columns (if present): “nCount_RNA”, “nFeature_RNA”,
“percent_mito”, “percent_ribo”, “percent_mito_ribo”.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
Median_nCount_RNA
</th>
<th style="text-align:right;">
Median_nFeature_RNA
</th>
<th style="text-align:right;">
Median_percent_mito
</th>
<th style="text-align:right;">
Median_percent_ribo
</th>
<th style="text-align:right;">
Median_percent_mito_ribo
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
sample1
</td>
<td style="text-align:right;">
2193
</td>
<td style="text-align:right;">
813.0
</td>
<td style="text-align:right;">
2.005731
</td>
<td style="text-align:right;">
36.91964
</td>
<td style="text-align:right;">
39.08969
</td>
</tr>
<tr>
<td style="text-align:left;">
sample2
</td>
<td style="text-align:right;">
2178
</td>
<td style="text-align:right;">
812.0
</td>
<td style="text-align:right;">
2.016383
</td>
<td style="text-align:right;">
37.03217
</td>
<td style="text-align:right;">
39.31227
</td>
</tr>
<tr>
<td style="text-align:left;">
sample3
</td>
<td style="text-align:right;">
2241
</td>
<td style="text-align:right;">
820.0
</td>
<td style="text-align:right;">
1.982099
</td>
<td style="text-align:right;">
37.40017
</td>
<td style="text-align:right;">
39.24930
</td>
</tr>
<tr>
<td style="text-align:left;">
sample4
</td>
<td style="text-align:right;">
2251
</td>
<td style="text-align:right;">
836.5
</td>
<td style="text-align:right;">
2.057405
</td>
<td style="text-align:right;">
36.16109
</td>
<td style="text-align:right;">
38.38636
</td>
</tr>
<tr>
<td style="text-align:left;">
Totals (All Cells)
</td>
<td style="text-align:right;">
2213
</td>
<td style="text-align:right;">
819.0
</td>
<td style="text-align:right;">
2.010702
</td>
<td style="text-align:right;">
36.92524
</td>
<td style="text-align:right;">
39.03276
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="additional-variables">Additional Variables<a class="anchor" aria-label="anchor" href="#additional-variables"></a>
</h4>
<p>In addition to default variables, users can supply their own
additional meta data columns to calculate medians using the
<code>median_var</code> parameter.<br><em>NOTE: The meta data column must be in numeric format (e.g., integer,
numeric, dbl)</em></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">median_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Median_Stats.html">Median_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span>, median_var <span class="op">=</span> <span class="st">"module_score1"</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
Median_nCount_RNA
</th>
<th style="text-align:right;">
Median_nFeature_RNA
</th>
<th style="text-align:right;">
Median_percent_mito
</th>
<th style="text-align:right;">
Median_percent_ribo
</th>
<th style="text-align:right;">
Median_percent_mito_ribo
</th>
<th style="text-align:right;">
Median_module_score1
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
sample1
</td>
<td style="text-align:right;">
2193
</td>
<td style="text-align:right;">
813.0
</td>
<td style="text-align:right;">
2.005731
</td>
<td style="text-align:right;">
36.91964
</td>
<td style="text-align:right;">
39.08969
</td>
<td style="text-align:right;">
-0.0545099
</td>
</tr>
<tr>
<td style="text-align:left;">
sample2
</td>
<td style="text-align:right;">
2178
</td>
<td style="text-align:right;">
812.0
</td>
<td style="text-align:right;">
2.016383
</td>
<td style="text-align:right;">
37.03217
</td>
<td style="text-align:right;">
39.31227
</td>
<td style="text-align:right;">
-0.0678532
</td>
</tr>
<tr>
<td style="text-align:left;">
sample3
</td>
<td style="text-align:right;">
2241
</td>
<td style="text-align:right;">
820.0
</td>
<td style="text-align:right;">
1.982099
</td>
<td style="text-align:right;">
37.40017
</td>
<td style="text-align:right;">
39.24930
</td>
<td style="text-align:right;">
-0.0793297
</td>
</tr>
<tr>
<td style="text-align:left;">
sample4
</td>
<td style="text-align:right;">
2251
</td>
<td style="text-align:right;">
836.5
</td>
<td style="text-align:right;">
2.057405
</td>
<td style="text-align:right;">
36.16109
</td>
<td style="text-align:right;">
38.38636
</td>
<td style="text-align:right;">
-0.1935054
</td>
</tr>
<tr>
<td style="text-align:left;">
Totals (All Cells)
</td>
<td style="text-align:right;">
2213
</td>
<td style="text-align:right;">
819.0
</td>
<td style="text-align:right;">
2.010702
</td>
<td style="text-align:right;">
36.92524
</td>
<td style="text-align:right;">
39.03276
</td>
<td style="text-align:right;">
-0.0922378
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="calculate-median-absolute-deviations">Calculate Median Absolute Deviations<a class="anchor" aria-label="anchor" href="#calculate-median-absolute-deviations"></a>
</h4>
<p>In addition to calculating the median values scCustomize includes
function to calculate the median absolute deviation for each of those
features with function <code>MAD_Stats</code>. By setting the parameter
<code>mad_num</code> the function will retrun the MAD*mad_num.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MAD_Stats.html">MAD_Stats</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"orig.ident"</span>, mad_num <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
orig.ident
</th>
<th style="text-align:right;">
MAD x 2 nCount_RNA
</th>
<th style="text-align:right;">
MAD x 2 nFeature_RNA
</th>
<th style="text-align:right;">
MAD x 2 percent_mito
</th>
<th style="text-align:right;">
MAD x 2 percent_ribo
</th>
<th style="text-align:right;">
MAD x 2 percent_mito_ribo
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
sample1
</td>
<td style="text-align:right;">
1464.809
</td>
<td style="text-align:right;">
397.3368
</td>
<td style="text-align:right;">
1.640639
</td>
<td style="text-align:right;">
23.92318
</td>
<td style="text-align:right;">
23.93974
</td>
</tr>
<tr>
<td style="text-align:left;">
sample2
</td>
<td style="text-align:right;">
1319.514
</td>
<td style="text-align:right;">
361.7544
</td>
<td style="text-align:right;">
1.579981
</td>
<td style="text-align:right;">
23.47474
</td>
<td style="text-align:right;">
23.15831
</td>
</tr>
<tr>
<td style="text-align:left;">
sample3
</td>
<td style="text-align:right;">
1500.391
</td>
<td style="text-align:right;">
392.8890
</td>
<td style="text-align:right;">
1.557364
</td>
<td style="text-align:right;">
23.57180
</td>
<td style="text-align:right;">
23.52481
</td>
</tr>
<tr>
<td style="text-align:left;">
sample4
</td>
<td style="text-align:right;">
1478.152
</td>
<td style="text-align:right;">
381.0282
</td>
<td style="text-align:right;">
1.451580
</td>
<td style="text-align:right;">
24.44613
</td>
<td style="text-align:right;">
24.08378
</td>
</tr>
<tr>
<td style="text-align:left;">
Totals (All Cells)
</td>
<td style="text-align:right;">
1442.570
</td>
<td style="text-align:right;">
379.5456
</td>
<td style="text-align:right;">
1.544386
</td>
<td style="text-align:right;">
23.73656
</td>
<td style="text-align:right;">
23.38683
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="plotting-median-data">Plotting Median Data<a class="anchor" aria-label="anchor" href="#plotting-median-data"></a>
</h3>
<p>scCustomize also contains series of functions for plotting the
results of these median calculations:</p>
<ul>
<li><code><a href="../reference/Plot_Median_Genes.html">Plot_Median_Genes()</a></code></li>
<li><code><a href="../reference/Plot_Median_UMIs.html">Plot_Median_UMIs()</a></code></li>
<li><code><a href="../reference/Plot_Median_Mito.html">Plot_Median_Mito()</a></code></li>
<li><code><a href="../reference/Plot_Median_Other.html">Plot_Median_Other()</a></code></li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Plot_Median_Genes.html">Plot_Median_Genes</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Plot_Median_Genes.html">Plot_Median_Genes</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/Plot_Median_Other.html">Plot_Median_Other</a></span><span class="op">(</span>seurat_object <span class="op">=</span> <span class="va">pbmc</span>, median_var <span class="op">=</span> <span class="st">"module_score1"</span>, group.by <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Statistics_files/figure-html/unnamed-chunk-27-1.png" width="1248" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
