<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Updating Gene Symbols • scCustomize</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Updating Gene Symbols">
<meta property="og:description" content="scCustomize">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
    <li>
      <a href="../articles/cNMF_Functionality.html">cNMF Functionality &amp; Plotting</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Update_Gene_Symbols_files/kePrint-0.0.1/kePrint.js"></script><link href="Update_Gene_Symbols_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Updating Gene Symbols</h1>
            
            <h4 data-toc-skip class="date">Compiled: September 10,
2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/vignettes/articles/Update_Gene_Symbols.Rmd" class="external-link"><code>vignettes/articles/Update_Gene_Symbols.Rmd</code></a></small>
      <div class="hidden name"><code>Update_Gene_Symbols.Rmd</code></div>

    </div>

    
    
<hr>
<style>
p.caption {
  font-size: 0.9em;
}
</style>
<div class="section level2">
<h2 id="upating-human-gene-symbols">Upating Human Gene Symbols<a class="anchor" aria-label="anchor" href="#upating-human-gene-symbols"></a>
</h2>
<p>The official gene symbols used in a dataset can change depending on
the reference version used in aligning that particular dataset. For
human genes the official symbols are set by <a href="https://www.genenames.org/" class="external-link">HGNC</a>.</p>
<p>In the absence of more static identifier (Ensembl ID or Entrez ID
Numbers) the only way to update gene symbols is to examine the current
and past symbols for all genes in the HGNC database. However, many of
the functions that perform this task come with caveats that vary from
lack of ease of updating to newest HGNC data or at worst potentially
improperly renaming symbols.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load Packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">scCustomize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/qsbase/qs" class="external-link">qs</a></span><span class="op">)</span></span></code></pre></div>
<p>Load Seurat Object &amp; Add QC Data</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read object</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">pbmc3k.SeuratData</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/pbmc3k.SeuratData/man/pbmc3k.final.html" class="external-link">pbmc3k.final</a></span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="issues-with-other-functions">Issues with other functions<a class="anchor" aria-label="anchor" href="#issues-with-other-functions"></a>
</h3>
<p>In order to understand how scCustomize’s
<code>Update_HGNC_Symbols()</code> improves process it is important to
be aware of the caveats of some other tools.</p>
<div class="section level4">
<h4 id="seurats-updatesymbollist">Seurat’s <code>UpdateSymbolList()</code><a class="anchor" aria-label="anchor" href="#seurats-updatesymbollist"></a>
</h4>
<p>The first is the Seurat’s <code><a href="https://satijalab.org/seurat/reference/UpdateSymbolList.html" class="external-link">UpdateSymbolList()</a></code> which takes
an input vector of symbols and uses active connection to HGNC to query
for updated symbols. However, there are two caveats with this function
1) it requires user to have internet connection anytime using the
function, 2) it can potentially rename symbols incorrectly.</p>
<p>To illustrate the second issue I will use 3 gene symbols that have
been current for some time: MCM2, MCM7, CCNL1. However, let’s take a
look at some of the previous symbols for each of these genes:<br>
- Previous Symbols for MCM2 are: CCNL1 &amp; CDCL1<br>
- Previous symbols for MCM7 are: MCM2<br>
- Previous symbols for CCNL1 are: None</p>
<p>Now see what happens when we use <code>UpdateSymbolList</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_symbols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MCM2"</span>, <span class="st">"MCM7"</span>, <span class="st">"CCNL1"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/UpdateSymbolList.html" class="external-link">UpdateSymbolList</a></span><span class="op">(</span>symbols <span class="op">=</span> <span class="va">test_symbols</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MCM7" "MCM7" "MCM2"</span></span></code></pre>
<p>As you can see the functions does the following:<br>
- Renames MCM2 &gt; MCM7 because MCM2 is a previous symbol.<br>
- Leaves MCM7 the same because no other gene has MCM7 as previous
symbol.<br>
- Renames CCNL1 &gt; MCM2 because CCNL1 is previous symbol.</p>
<p>The reason that this happens is because <code>UpdateSymbolList</code>
queries each symbol in isolation and not in the context of all of the
genes being queried.</p>
</div>
<div class="section level4">
<h4 id="hgnchelper-package">HGNChelper Package<a class="anchor" aria-label="anchor" href="#hgnchelper-package"></a>
</h4>
<p>After developing this function I was made aware of the HGNChelper
package which also aims to provide symbol updates. It solves renaming
issue in similar fashion to scCustomize (see below). It also provides a
solution for requirement of internet access.</p>
<p>It does this by storing HGNC dataset as package data so that it comes
bundled with the package. However, there is an issue with the way this
is implemented. First, the bundled data is from 2019 so is approached 5
years old. Updated data can be downloaded interactively using a package
function but this must be done in every R session where the data is
needed requiring internet access to use current data. The authors do
provide a solution to this but it involves cloning the github repo and
running source scripts which may be beyond many R users.</p>
</div>
</div>
<div class="section level3">
<h3 id="solving-the-issue-with-sccustomizes-update_hgnc_symbols">Solving the Issue with scCustomize’s
<code>Update_HGNC_Symbols</code><a class="anchor" aria-label="anchor" href="#solving-the-issue-with-sccustomizes-update_hgnc_symbols"></a>
</h3>
<p>scCustomize now provides the function
<code>Update_HGNC_Symbols</code> to attempt to solve both of these
caveats.</p>
<div class="section level4">
<h4 id="requirement-of-internet-access">Requirement of internet access<a class="anchor" aria-label="anchor" href="#requirement-of-internet-access"></a>
</h4>
<p><code>Update_HGNC_Symbols</code> does require internet access the
first time the function is being used to download most recent data from
HGNC. However, it then stores the downloaded data using <a href="https://bioconductor.org/packages/release/bioc/html/BiocFileCache.html" class="external-link">BiocFileCache
package</a>, meaning subsequent uses don’t require any internet access.
This also significant improves the speed of the function.</p>
</div>
<div class="section level4">
<h4 id="inappropriate-renaming">Inappropriate renaming<a class="anchor" aria-label="anchor" href="#inappropriate-renaming"></a>
</h4>
<p>Second, <code>Update_HGNC_Symbols</code> uses the full input list and
first automatically approves any symbol that is already an approved gene
symbol so that there is not a chance of improperly updating any symbols.
It then checks the remaining symbols for any symbol updates.</p>
<p>Let’s run our test symbol set:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Updated_HGNC_Symbols.html">Updated_HGNC_Symbols</a></span><span class="op">(</span>input_data <span class="op">=</span> <span class="va">test_symbols</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Input features contained <span style="color: #00BB00;">3</span> gene symbols</span></span>
<span><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #00BB00;">3</span> were already approved symbols.</span></span>
<span><span class="co">## <span style="color: #0000BB;">→</span> <span style="color: #00BB00;">0</span> were updated to approved symbol.</span></span>
<span><span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #00BB00;">0</span> were not found in HGNC dataset and remain unchanged.</span></span></code></pre>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
input_features
</th>
<th style="text-align:left;">
Approved_Symbol
</th>
<th style="text-align:left;">
Not_Found_Symbol
</th>
<th style="text-align:left;">
Updated_Symbol
</th>
<th style="text-align:left;">
Output_Features
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
MCM2
</td>
<td style="text-align:left;">
MCM2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
MCM2
</td>
</tr>
<tr>
<td style="text-align:left;">
MCM7
</td>
<td style="text-align:left;">
MCM7
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
MCM7
</td>
</tr>
<tr>
<td style="text-align:left;">
CCNL1
</td>
<td style="text-align:left;">
CCNL1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
CCNL1
</td>
</tr>
</tbody>
</table>
<p>As mentioned before the function is also very quick. Returning
updated symbols for 36,000 genes in ~1 second.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in full 10X reference genome feature list</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="st">"assets/Barcode_Rank_Example/sample1/outs/raw_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load tictoc to give timing</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jabiru/tictoc" class="external-link">tictoc</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get updated symbols</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">tic</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Updated_HGNC_Symbols.html">Updated_HGNC_Symbols</a></span><span class="op">(</span>input_data <span class="op">=</span> <span class="va">features</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Input features contained <span style="color: #00BB00;">36,601</span> gene symbols</span></span>
<span><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #00BB00;">23,196</span> were already approved symbols.</span></span>
<span><span class="co">## <span style="color: #0000BB;">→</span> <span style="color: #00BB00;">813</span> were updated to approved symbol.</span></span>
<span><span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #00BB00;">12,594</span> were not found in HGNC dataset and remain unchanged.</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/tictoc/man/tic.html" class="external-link">toc</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 0.323 sec elapsed</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="examining-the-results">Examining the Results<a class="anchor" aria-label="anchor" href="#examining-the-results"></a>
</h4>
<p>Now let’s take a look at the output from
<code>Updated_HGNC_Symbols</code>, which also has some detail advtanages
vs other methods.</p>
<p>For this example I have picked section of the results that contains
all 3 potential results.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span><span class="op">[</span><span class="fl">168</span><span class="op">:</span><span class="fl">177</span>, <span class="op">]</span></span></code></pre></div>
<table class="table table table-bordered table-condensed table-responsive table-striped" style="margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
input_features
</th>
<th style="text-align:left;">
Approved_Symbol
</th>
<th style="text-align:left;">
Not_Found_Symbol
</th>
<th style="text-align:left;">
Updated_Symbol
</th>
<th style="text-align:left;">
Output_Features
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
168
</td>
<td style="text-align:left;">
NPHP4
</td>
<td style="text-align:left;">
NPHP4
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NPHP4
</td>
</tr>
<tr>
<td style="text-align:left;">
169
</td>
<td style="text-align:left;">
KCNAB2
</td>
<td style="text-align:left;">
KCNAB2
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
KCNAB2
</td>
</tr>
<tr>
<td style="text-align:left;">
170
</td>
<td style="text-align:left;">
CHD5
</td>
<td style="text-align:left;">
CHD5
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
CHD5
</td>
</tr>
<tr>
<td style="text-align:left;">
171
</td>
<td style="text-align:left;">
RPL22
</td>
<td style="text-align:left;">
RPL22
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
RPL22
</td>
</tr>
<tr>
<td style="text-align:left;">
172
</td>
<td style="text-align:left;">
AL031847.1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
AL031847.1
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
AL031847.1
</td>
</tr>
<tr>
<td style="text-align:left;">
173
</td>
<td style="text-align:left;">
RNF207
</td>
<td style="text-align:left;">
RNF207
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
RNF207
</td>
</tr>
<tr>
<td style="text-align:left;">
174
</td>
<td style="text-align:left;">
ICMT
</td>
<td style="text-align:left;">
ICMT
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
ICMT
</td>
</tr>
<tr>
<td style="text-align:left;">
175
</td>
<td style="text-align:left;">
LINC00337
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
ICMT-DT
</td>
<td style="text-align:left;">
ICMT-DT
</td>
</tr>
<tr>
<td style="text-align:left;">
176
</td>
<td style="text-align:left;">
HES3
</td>
<td style="text-align:left;">
HES3
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
HES3
</td>
</tr>
<tr>
<td style="text-align:left;">
177
</td>
<td style="text-align:left;">
GPR153
</td>
<td style="text-align:left;">
GPR153
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
NA
</td>
<td style="text-align:left;">
GPR153
</td>
</tr>
</tbody>
</table>
<p>As you can see the majority of these symbols are already updated so
the input symbol matches the output symbol.</p>
<p>In the case of “AL031847.1” that annotation was not found in HGNC and
therefore the symbol was left unchanged.</p>
<p>Finally in the case of “LINC00337” there was an updated symbol of
“ICMT-DT” so the output symbol was updated to that current symbol.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="updating-mouse-gene-symbols">Updating Mouse Gene Symbols<a class="anchor" aria-label="anchor" href="#updating-mouse-gene-symbols"></a>
</h2>
<p>scCustomize also contains helper function to update mouse gene
symbols using the <a href="https://www.informatics.jax.org/" class="external-link">Mouse
Genome Informatics (MGI)</a> database.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load mouse dataset</span></span>
<span><span class="va">marsh_mouse_micro</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"assets/marsh_2020_micro.qs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get updated symbols</span></span>
<span><span class="va">updated_symbols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Updated_MGI_Symbols.html">Updated_MGI_Symbols</a></span><span class="op">(</span>input_data <span class="op">=</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Features</a></span><span class="op">(</span><span class="va">marsh_mouse_micro</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Input features contained <span style="color: #00BB00;">12,674</span> gene symbols</span></span>
<span><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #00BB00;">11,983</span> were already approved symbols.</span></span>
<span><span class="co">## <span style="color: #0000BB;">→</span> <span style="color: #00BB00;">541</span> were updated to approved symbol.</span></span>
<span><span class="co">## <span style="color: #BB0000;">✖</span> <span style="color: #00BB00;">152</span> were not found in MGI dataset and remain unchanged.</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
