<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ scCustomize</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="read-write-data">Read &amp; Write Data <a href="#read-write-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for reading and writing single cell sequencing data from raw data files.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Create_10X_H5.html">Create_10X_H5()</a></code> </p>
        </td>
        <td><p>Create H5 from 10X Outputs</p></td>
      </tr><tr><td>
          <p><code><a href="Create_CellBender_Merged_Seurat.html">Create_CellBender_Merged_Seurat()</a></code> </p>
        </td>
        <td><p>Create Seurat Object with Cell Bender and Raw data</p></td>
      </tr><tr><td>
          <p><code><a href="Pull_Directory_List.html">Pull_Directory_List()</a></code> </p>
        </td>
        <td><p>Pull Directory List</p></td>
      </tr><tr><td>
          <p><code><a href="Read10X_GEO.html">Read10X_GEO()</a></code> </p>
        </td>
        <td><p>Load in NCBI GEO data from 10X</p></td>
      </tr><tr><td>
          <p><code><a href="Read10X_Multi_Directory.html">Read10X_Multi_Directory()</a></code> </p>
        </td>
        <td><p>Load 10X count matrices from multiple directories</p></td>
      </tr><tr><td>
          <p><code><a href="Read10X_h5_GEO.html">Read10X_h5_GEO()</a></code> </p>
        </td>
        <td><p>Load in NCBI GEO data from 10X in HDF5 file format</p></td>
      </tr><tr><td>
          <p><code><a href="Read10X_h5_Multi_Directory.html">Read10X_h5_Multi_Directory()</a></code> </p>
        </td>
        <td><p>Load 10X h5 count matrices from multiple directories</p></td>
      </tr><tr><td>
          <p><code><a href="Read_CellBender_h5_Mat.html">Read_CellBender_h5_Mat()</a></code> </p>
        </td>
        <td><p>Load CellBender h5 matrices (corrected)</p></td>
      </tr><tr><td>
          <p><code><a href="Read_CellBender_h5_Multi_Directory.html">Read_CellBender_h5_Multi_Directory()</a></code> </p>
        </td>
        <td><p>Load CellBender h5 matrices (corrected) from multiple directories</p></td>
      </tr><tr><td>
          <p><code><a href="Read_CellBender_h5_Multi_File.html">Read_CellBender_h5_Multi_File()</a></code> </p>
        </td>
        <td><p>Load CellBender h5 matrices (corrected) from multiple files</p></td>
      </tr><tr><td>
          <p><code><a href="Read_GEO_Delim.html">Read_GEO_Delim()</a></code> </p>
        </td>
        <td><p>Load in NCBI GEO data formatted as single file per sample</p></td>
      </tr><tr><td>
          <p><code><a href="Read_Metrics_10X.html">Read_Metrics_10X()</a></code> </p>
        </td>
        <td><p>Read Overall Statistics from 10X Cell Ranger Count</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="sequencing-metrics-qc-plots">Sequencing Metrics QC Plots <a href="#sequencing-metrics-qc-plots" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions plotting various QC metrics provides by 10X Genomics Cell Ranger Outputs</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Seq_QC_Plot_Alignment_Combined.html">Seq_QC_Plot_Alignment_Combined()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment) (Layout)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Basic_Combined.html">Seq_QC_Plot_Basic_Combined()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Layout)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Genes.html">Seq_QC_Plot_Genes()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Number_Cells.html">Seq_QC_Plot_Number_Cells()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Reads_in_Cells.html">Seq_QC_Plot_Reads_in_Cells()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Reads_per_Cell.html">Seq_QC_Plot_Reads_per_Cell()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Saturation.html">Seq_QC_Plot_Saturation()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Total_Genes.html">Seq_QC_Plot_Total_Genes()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_UMIs.html">Seq_QC_Plot_UMIs()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Antisense.html">Seq_QC_Plot_Antisense()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Exonic.html">Seq_QC_Plot_Exonic()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Genome.html">Seq_QC_Plot_Genome()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Intergenic.html">Seq_QC_Plot_Intergenic()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Intronic.html">Seq_QC_Plot_Intronic()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment)</p></td>
      </tr><tr><td>
          <p><code><a href="Seq_QC_Plot_Transcriptome.html">Seq_QC_Plot_Transcriptome()</a></code> </p>
        </td>
        <td><p>QC Plots Sequencing metrics (Alignment)</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="basic-analysis-qc-plots">Basic Analysis QC Plots <a href="#basic-analysis-qc-plots" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions plotting various QC from Seurat Objects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="QC_Plot_GenevsFeature.html">QC_Plot_GenevsFeature()</a></code> </p>
        </td>
        <td><p>QC Plots Genes vs Misc</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plot_UMIvsFeature.html">QC_Plot_UMIvsFeature()</a></code> </p>
        </td>
        <td><p>QC Plots UMI vs Misc</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plot_UMIvsGene.html">QC_Plot_UMIvsGene()</a></code> </p>
        </td>
        <td><p>QC Plots Genes vs UMIs</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plots_Combined_Vln.html">QC_Plots_Combined_Vln()</a></code> </p>
        </td>
        <td><p>QC Plots Genes, UMIs, &amp; % Mito</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plots_Complexity.html">QC_Plots_Complexity()</a></code> </p>
        </td>
        <td><p>QC Plots Cell "Complexity"</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plots_Feature.html">QC_Plots_Feature()</a></code> </p>
        </td>
        <td><p>QC Plots Feature</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plots_Genes.html">QC_Plots_Genes()</a></code> </p>
        </td>
        <td><p>QC Plots Genes</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plots_Mito.html">QC_Plots_Mito()</a></code> </p>
        </td>
        <td><p>QC Plots Mito</p></td>
      </tr><tr><td>
          <p><code><a href="QC_Plots_UMIs.html">QC_Plots_UMIs()</a></code> </p>
        </td>
        <td><p>QC Plots UMIs</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="seurat-plotting-functions">Seurat Plotting Functions <a href="#seurat-plotting-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions customization and plotting of single cell data/results from Seurat Objects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Cell_Highlight_Plot.html">Cell_Highlight_Plot()</a></code> </p>
        </td>
        <td><p>Meta Highlight Plot</p></td>
      </tr><tr><td>
          <p><code><a href="Cluster_Highlight_Plot.html">Cluster_Highlight_Plot()</a></code> </p>
        </td>
        <td><p>Cluster Highlight Plot</p></td>
      </tr><tr><td>
          <p><code><a href="Clustered_DotPlot.html">Clustered_DotPlot()</a></code> </p>
        </td>
        <td><p>Clustered DotPlot</p></td>
      </tr><tr><td>
          <p><code><a href="DimPlot_All_Samples.html">DimPlot_All_Samples()</a></code> </p>
        </td>
        <td><p>DimPlot by Meta Data Column</p></td>
      </tr><tr><td>
          <p><code><a href="DimPlot_scCustom.html">DimPlot_scCustom()</a></code> </p>
        </td>
        <td><p>DimPlot with modified default settings</p></td>
      </tr><tr><td>
          <p><code><a href="DotPlot_scCustom.html">DotPlot_scCustom()</a></code> </p>
        </td>
        <td><p>Customized DotPlot</p></td>
      </tr><tr><td>
          <p><code><a href="FeaturePlot_DualAssay.html">FeaturePlot_DualAssay()</a></code> </p>
        </td>
        <td><p>Customize FeaturePlot of two assays</p></td>
      </tr><tr><td>
          <p><code><a href="FeaturePlot_scCustom.html">FeaturePlot_scCustom()</a></code> </p>
        </td>
        <td><p>Customize FeaturePlot</p></td>
      </tr><tr><td>
          <p><code><a href="Meta_Highlight_Plot.html">Meta_Highlight_Plot()</a></code> </p>
        </td>
        <td><p>Meta Highlight Plot</p></td>
      </tr><tr><td>
          <p><code><a href="PC_Plotting.html">PC_Plotting()</a></code> </p>
        </td>
        <td><p>PC Plots</p></td>
      </tr><tr><td>
          <p><code><a href="Split_FeatureScatter.html">Split_FeatureScatter()</a></code> </p>
        </td>
        <td><p>Split FeatureScatter</p></td>
      </tr><tr><td>
          <p><code><a href="Stacked_VlnPlot.html">Stacked_VlnPlot()</a></code> </p>
        </td>
        <td><p>Stacked Violin Plot</p></td>
      </tr><tr><td>
          <p><code><a href="VariableFeaturePlot_scCustom.html">VariableFeaturePlot_scCustom()</a></code> </p>
        </td>
        <td><p>Custom Labeled Variable Features Plot</p></td>
      </tr><tr><td>
          <p><code><a href="VlnPlot_scCustom.html">VlnPlot_scCustom()</a></code> </p>
        </td>
        <td><p>VlnPlot with modified default settings</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Density_Custom.html">Plot_Density_Custom()</a></code> </p>
        </td>
        <td><p>Nebulosa Density Plot</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Density_Joint_Only.html">Plot_Density_Joint_Only()</a></code> </p>
        </td>
        <td><p>Nebulosa Joint Density Plot</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="iterative-plotting-functions">Iterative Plotting Functions <a href="#iterative-plotting-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions for quick iterative plotting.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Iterate_Cluster_Highlight_Plot.html">Iterate_Cluster_Highlight_Plot()</a></code> </p>
        </td>
        <td><p>Iterate Cluster Highlight Plot</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_DimPlot_bySample.html">Iterate_DimPlot_bySample()</a></code> </p>
        </td>
        <td><p>Iterate DimPlot By Sample</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_FeaturePlot_scCustom.html">Iterate_FeaturePlot_scCustom()</a></code> </p>
        </td>
        <td><p>Iterative Plotting of Gene Lists using Custom FeaturePlots</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_Meta_Highlight_Plot.html">Iterate_Meta_Highlight_Plot()</a></code> </p>
        </td>
        <td><p>Iterate Meta Highlight Plot</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_PC_Loading_Plots.html">Iterate_PC_Loading_Plots()</a></code> </p>
        </td>
        <td><p>Iterate PC Loading Plots</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_Plot_Density_Custom.html">Iterate_Plot_Density_Custom()</a></code> </p>
        </td>
        <td><p>Iterative Plotting of Gene Lists using Custom Density Plots</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_Plot_Density_Joint.html">Iterate_Plot_Density_Joint()</a></code> </p>
        </td>
        <td><p>Iterative Plotting of Gene Lists using Custom Joint Density Plots</p></td>
      </tr><tr><td>
          <p><code><a href="Iterate_VlnPlot_scCustom.html">Iterate_VlnPlot_scCustom()</a></code> </p>
        </td>
        <td><p>Iterative Plotting of Gene Lists using VlnPlot_scCustom</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="liger-plotting-functions">LIGER Plotting Functions <a href="#liger-plotting-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions customization and plotting of single cell data/results from LIGER Objects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="DimPlot_LIGER.html">DimPlot_LIGER()</a></code> </p>
        </td>
        <td><p>DimPlot LIGER Version</p></td>
      </tr><tr><td>
          <p><code><a href="plotFactors_scCustom.html">plotFactors_scCustom()</a></code> </p>
        </td>
        <td><p>Customized version of plotFactors</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="plotting-utilities-color-palettes-">Plotting Utilities #1 (Color Palettes) <a href="#plotting-utilities-color-palettes-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Helper functions for quick access to useful or default color palettes.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ColorBlind_Pal.html">ColorBlind_Pal()</a></code> </p>
        </td>
        <td><p>Color Universal Design Short Palette</p></td>
      </tr><tr><td>
          <p><code><a href="Dark2_Pal.html">Dark2_Pal()</a></code> </p>
        </td>
        <td><p>Dark2 Palette</p></td>
      </tr><tr><td>
          <p><code><a href="DiscretePalette_scCustomize.html">DiscretePalette_scCustomize()</a></code> </p>
        </td>
        <td><p>Discrete color palettes</p></td>
      </tr><tr><td>
          <p><code><a href="Hue_Pal.html">Hue_Pal()</a></code> </p>
        </td>
        <td><p>Hue_Pal</p></td>
      </tr><tr><td>
          <p><code><a href="JCO_Four.html">JCO_Four()</a></code> </p>
        </td>
        <td><p>Four Color Palette (JCO)</p></td>
      </tr><tr><td>
          <p><code><a href="NavyAndOrange.html">NavyAndOrange()</a></code> </p>
        </td>
        <td><p>Navy and Orange Dual Color Palette</p></td>
      </tr><tr><td>
          <p><code><a href="PalettePlot.html">PalettePlot()</a></code> </p>
        </td>
        <td><p>Plot color palette in viewer</p></td>
      </tr><tr><td>
          <p><code><a href="Single_Color_Palette.html">Single_Color_Palette()</a></code> </p>
        </td>
        <td><p>Single Color Palettes for Plotting</p></td>
      </tr><tr><td>
          <p><code><a href="scCustomize_Palette.html">scCustomize_Palette()</a></code> </p>
        </td>
        <td><p>Color Palette Selection for scCustomize</p></td>
      </tr><tr><td>
          <p><code><a href="viridis_shortcut.html">viridis_plasma_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_plasma_light_high</a></code> <code><a href="viridis_shortcut.html">viridis_inferno_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_inferno_light_high</a></code> <code><a href="viridis_shortcut.html">viridis_magma_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_magma_light_high</a></code> <code><a href="viridis_shortcut.html">viridis_dark_high</a></code> <code><a href="viridis_shortcut.html">viridis_light_high</a></code> </p>
        </td>
        <td><p>Viridis Shortcuts</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="plotting-utilities-themes-theme-adjustment-">Plotting Utilities #2 (Themes/Theme Adjustment) <a href="#plotting-utilities-themes-theme-adjustment-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Helper functions for quick modification of theme elements and custom themes.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Blank_Theme.html">Blank_Theme()</a></code> </p>
        </td>
        <td><p>Blank Theme</p></td>
      </tr><tr><td>
          <p><code><a href="Move_Legend.html">Move_Legend()</a></code> </p>
        </td>
        <td><p>Move Legend Position</p></td>
      </tr><tr><td>
          <p><code><a href="UnRotate_X.html">UnRotate_X()</a></code> </p>
        </td>
        <td><p>Unrotate x axis on VlnPlot</p></td>
      </tr><tr><td>
          <p><code><a href="theme_ggprism_mod.html">theme_ggprism_mod()</a></code> </p>
        </td>
        <td><p>Modified ggprism theme</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="helper-utilities-seurat-">Helper Utilities (Seurat) <a href="#helper-utilities-seurat-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to provide ease of use for frequently used code from Seurat Objects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Case_Check.html">Case_Check()</a></code> </p>
        </td>
        <td><p>Check for alternate case features
Checks Seurat object for the presence of features with the same spelling but alternate case.</p></td>
      </tr><tr><td>
          <p><code><a href="Change_Delim_All.html">Change_Delim_All()</a></code> </p>
        </td>
        <td><p>Change all delimiters in cell name</p></td>
      </tr><tr><td>
          <p><code><a href="Change_Delim_Prefix.html">Change_Delim_Prefix()</a></code> </p>
        </td>
        <td><p>Change barcode prefix delimiter</p></td>
      </tr><tr><td>
          <p><code><a href="Change_Delim_Suffix.html">Change_Delim_Suffix()</a></code> </p>
        </td>
        <td><p>Change barcode suffix delimiter</p></td>
      </tr><tr><td>
          <p><code><a href="CheckMatrix_scCustom.html">CheckMatrix_scCustom()</a></code> </p>
        </td>
        <td><p>Check Matrix Validity</p></td>
      </tr><tr><td>
          <p><code><a href="Extract_Modality.html">Extract_Modality()</a></code> </p>
        </td>
        <td><p>Extract multi-modal data into list by modality</p></td>
      </tr><tr><td>
          <p><code><a href="Fetch_Meta.html">Fetch_Meta()</a></code> </p>
        </td>
        <td><p>Get meta data from object</p></td>
      </tr><tr><td>
          <p><code><a href="Gene_Present.html">Gene_Present()</a></code> </p>
        </td>
        <td><p>Check if genes/features are present</p></td>
      </tr><tr><td>
          <p><code><a href="Merge_Sparse_Data_All.html">Merge_Sparse_Data_All()</a></code> </p>
        </td>
        <td><p>Merge a list of Sparse Matrices</p></td>
      </tr><tr><td>
          <p><code><a href="Merge_Sparse_Multimodal_All.html">Merge_Sparse_Multimodal_All()</a></code> </p>
        </td>
        <td><p>Merge a list of Sparse Matrices contain multi-modal data.</p></td>
      </tr><tr><td>
          <p><code><a href="Meta_Numeric.html">Meta_Numeric()</a></code> </p>
        </td>
        <td><p>Check if meta data columns are numeric</p></td>
      </tr><tr><td>
          <p><code><a href="Meta_Present.html">Meta_Present()</a></code> </p>
        </td>
        <td><p>Check if meta data are present</p></td>
      </tr><tr><td>
          <p><code><a href="Reduction_Loading_Present.html">Reduction_Loading_Present()</a></code> </p>
        </td>
        <td><p>Check if reduction loadings are present</p></td>
      </tr><tr><td>
          <p><code><a href="Replace_Suffix.html">Replace_Suffix()</a></code> </p>
        </td>
        <td><p>Replace barcode suffixes</p></td>
      </tr><tr><td>
          <p><code><a href="Add_CellBender_Diff.html">Add_CellBender_Diff()</a></code> </p>
        </td>
        <td><p>Calculate and add differences post-cell bender analysis</p></td>
      </tr><tr><td>
          <p><code><a href="Add_Cell_Complexity_Seurat.html">Add_Cell_Complexity_Seurat()</a></code> </p>
        </td>
        <td><p>Add Cell Complexity Value</p></td>
      </tr><tr><td>
          <p><code><a href="Add_Mito_Ribo_Seurat.html">Add_Mito_Ribo_Seurat()</a></code> </p>
        </td>
        <td><p>Add Mito and Ribo percentages</p></td>
      </tr><tr><td>
          <p><code><a href="Add_Sample_Meta.html">Add_Sample_Meta()</a></code> </p>
        </td>
        <td><p>Add Sample Level Meta Data</p></td>
      </tr><tr><td>
          <p><code><a href="Extract_Sample_Meta.html">Extract_Sample_Meta()</a></code> </p>
        </td>
        <td><p>Extract sample level meta.data</p></td>
      </tr><tr><td>
          <p><code><a href="Liger_to_Seurat.html">Liger_to_Seurat()</a></code> </p>
        </td>
        <td><p>Create a Seurat object containing the data from a liger object</p></td>
      </tr><tr><td>
          <p><code><a href="Merge_Seurat_List.html">Merge_Seurat_List()</a></code> </p>
        </td>
        <td><p>Merge a list of Seurat Objects</p></td>
      </tr><tr><td>
          <p><code><a href="Meta_Remove_Seurat.html">Meta_Remove_Seurat()</a></code> </p>
        </td>
        <td><p>Remove meta data columns containing Seurat Defaults</p></td>
      </tr><tr><td>
          <p><code><a href="Rename_Clusters.html">Rename_Clusters()</a></code> </p>
        </td>
        <td><p>Rename Cluster Seurat</p></td>
      </tr><tr><td>
          <p><code><a href="Store_Misc_Info_Seurat.html">Store_Misc_Info_Seurat()</a></code> </p>
        </td>
        <td><p>Store misc data in Seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="Store_Palette_Seurat.html">Store_Palette_Seurat()</a></code> </p>
        </td>
        <td><p>Store color palette in Seurat object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="helper-utilities-liger-">Helper Utilities (LIGER) <a href="#helper-utilities-liger-" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to provide ease of use for frequently used code from LIGER Objects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Add_Cell_Complexity_LIGER.html">Add_Cell_Complexity_LIGER()</a></code> </p>
        </td>
        <td><p>Add Cell Complexity Value</p></td>
      </tr><tr><td>
          <p><code><a href="Add_Mito_Ribo_LIGER.html">Add_Mito_Ribo_LIGER()</a></code> </p>
        </td>
        <td><p>Add Mito and Ribo percentages to LIGER</p></td>
      </tr><tr><td>
          <p><code><a href="Meta_Present_LIGER.html">Meta_Present_LIGER()</a></code> </p>
        </td>
        <td><p>Check if meta data are present</p></td>
      </tr><tr><td>
          <p><code><a href="Top_Genes_Factor.html">Top_Genes_Factor()</a></code> </p>
        </td>
        <td><p>Extract top loading genes for LIGER factor</p></td>
      </tr><tr><td>
          <p><code><a href="Variable_Features_ALL_LIGER.html">Variable_Features_ALL_LIGER()</a></code> </p>
        </td>
        <td><p>Perform variable gene selection over whole dataset</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="cluster-marker-annotation-utilities">Cluster/Marker Annotation Utilities <a href="#cluster-marker-annotation-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to provide ease of use or add functionality to DEG and cluster annotation.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Add_Pct_Diff.html">Add_Pct_Diff()</a></code> </p>
        </td>
        <td><p>Add percentage difference to DE results</p></td>
      </tr><tr><td>
          <p><code><a href="Create_Cluster_Annotation_File.html">Create_Cluster_Annotation_File()</a></code> </p>
        </td>
        <td><p>Create cluster annotation csv file</p></td>
      </tr><tr><td>
          <p><code><a href="Extract_Top_Markers.html">Extract_Top_Markers()</a></code> </p>
        </td>
        <td><p>Extract Top N Marker Genes</p></td>
      </tr><tr><td>
          <p><code><a href="Pull_Cluster_Annotation.html">Pull_Cluster_Annotation()</a></code> </p>
        </td>
        <td><p>Pull cluster information from annotation csv file.</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="project-organization-utilities">Project Organization Utilities <a href="#project-organization-utilities" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to provide ease of use for organization of analysis projects.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Copy_From_GCP.html">Copy_From_GCP()</a></code> </p>
        </td>
        <td><p>Copy folder from GCP bucket from R Console</p></td>
      </tr><tr><td>
          <p><code><a href="Copy_To_GCP.html">Copy_To_GCP()</a></code> </p>
        </td>
        <td><p>Copy folder to GCP bucket from R Console</p></td>
      </tr><tr><td>
          <p><code><a href="Setup_scRNAseq_Project.html">Setup_scRNAseq_Project()</a></code> </p>
        </td>
        <td><p>Setup project directory structure</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="statistics-functions">Statistics Functions <a href="#statistics-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions quick return of different object and data metrics.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="CellBender_Feature_Diff.html">CellBender_Feature_Diff()</a></code> </p>
        </td>
        <td><p>CellBender Feature Differences</p></td>
      </tr><tr><td>
          <p><code><a href="Cluster_Stats_All_Samples.html">Cluster_Stats_All_Samples()</a></code> </p>
        </td>
        <td><p>Calculate Cluster Stats</p></td>
      </tr><tr><td>
          <p><code><a href="Median_Stats.html">Median_Stats()</a></code> </p>
        </td>
        <td><p>Median Statistics</p></td>
      </tr><tr><td>
          <p><code><a href="Percent_Expressing.html">Percent_Expressing()</a></code> </p>
        </td>
        <td><p>Calculate percent of expressing cells</p></td>
      </tr><tr><td>
          <p><code><a href="CellBender_Diff_Plot.html">CellBender_Diff_Plot()</a></code> </p>
        </td>
        <td><p>Plot Number of Cells/Nuclei per Sample</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Cells_per_Sample.html">Plot_Cells_per_Sample()</a></code> </p>
        </td>
        <td><p>Plot Number of Cells/Nuclei per Sample</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Median_Genes.html">Plot_Median_Genes()</a></code> </p>
        </td>
        <td><p>Plot Median Genes per Cell per Sample</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Median_Mito.html">Plot_Median_Mito()</a></code> </p>
        </td>
        <td><p>Plot Median Percent Mito per Cell per Sample</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Median_Other.html">Plot_Median_Other()</a></code> </p>
        </td>
        <td><p>Plot Median other variable per Cell per Sample</p></td>
      </tr><tr><td>
          <p><code><a href="Plot_Median_UMIs.html">Plot_Median_UMIs()</a></code> </p>
        </td>
        <td><p>Plot Median UMIs per Cell per Sample</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="data">Data <a href="#data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Reference data for package functions.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ensembl_mito_id.html">ensembl_mito_id</a></code> </p>
        </td>
        <td><p>Ensembl Mito IDs</p></td>
      </tr><tr><td>
          <p><code><a href="ensembl_ribo_id.html">ensembl_ribo_id</a></code> </p>
        </td>
        <td><p>Ensembl Ribo IDs</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

