<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>scCustomize: Custom Visualizations &amp; Functions for Streamlined Analyses of Single Cell Sequencing — scCustomize-package • scCustomize</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="scCustomize: Custom Visualizations &amp; Functions for Streamlined Analyses of Single Cell Sequencing — scCustomize-package"><meta property="og:description" content="Collection of functions created and/or curated to aid in the visualization and analysis of single-cell data using 'R'. 'scCustomize' aims to provide 1) Customized visualizations for aid in ease of use and to create more aesthetic and functional visuals. 2) Improve speed/reproducibility of common tasks/pieces of code in scRNA-seq analysis with a single or group of functions. For citation please use: Marsh SE (2021) &quot;Custom Visualizations &amp;amp; Functions for Streamlined Analyses of Single Cell Sequencing&quot; doi:10.5281/zenodo.5706430
 RRID:SCR_024675."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scCustomize</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">3.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Getting Started</a>
</li>
<li>
  <a href="../articles/Installation.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Plotting Vignettes</li>
    <li>
      <a href="../articles/Gene_Expression_Plotting.html">Plotting #1: General Analysis Plots</a>
    </li>
    <li>
      <a href="../articles/QC_Plots.html">Plotting #2: QC Plots &amp; Analysis</a>
    </li>
    <li>
      <a href="../articles/Sequencing_QC_Plots.html">Plotting #3: Sequencing QC Plots/Analysis</a>
    </li>
    <li>
      <a href="../articles/Iterative_Plotting.html">Plotting #4: Iterative Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Spatial_Plotting.html">Plotting #5: Spatial Plotting Functions</a>
    </li>
    <li>
      <a href="../articles/Color_Palettes.html">Customized Color Palettes &amp; Themes</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Helpers &amp; Utilities</li>
    <li>
      <a href="../articles/Helpers_and_Utilities.html">General Helpers &amp; Utilities</a>
    </li>
    <li>
      <a href="../articles/Object_QC_Functions.html">Object QC Functions</a>
    </li>
    <li>
      <a href="../articles/Read_and_Write_Functions.html">Read &amp; Write Data Functions</a>
    </li>
    <li>
      <a href="../articles/Object_Conversion.html">Object/Assay Format Conversion</a>
    </li>
    <li>
      <a href="../articles/Markers_and_Cluster_Annotation.html">Marker Identification &amp; Cluster Annotation Helpers</a>
    </li>
    <li>
      <a href="../articles/Statistics.html">Statistics Functions</a>
    </li>
    <li>
      <a href="../articles/Update_Gene_Symbols.html">Updating Gene Symbols</a>
    </li>
    <li>
      <a href="../articles/Misc_Functions.html">Misc Functions</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Other Package Interactivity/Functions</li>
    <li>
      <a href="../articles/LIGER_Functions.html">LIGER Plotting &amp; Functionality</a>
    </li>
    <li>
      <a href="../articles/Cell_Bender_Functions.html">Cell Bender Functionality &amp; Plotting</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">News/Changelog</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/FAQ.html">FAQ</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/samuel-marsh/scCustomize/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>scCustomize: Custom Visualizations &amp; Functions for Streamlined Analyses of Single Cell Sequencing</h1>
    <small class="dont-index">Source: <a href="https://github.com/samuel-marsh/scCustomize/blob/HEAD/R/zzz.R" class="external-link"><code>R/zzz.R</code></a></small>
    <div class="hidden name"><code>scCustomize-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Collection of functions created and/or curated to aid in the visualization and analysis of single-cell data using 'R'. 'scCustomize' aims to provide 1) Customized visualizations for aid in ease of use and to create more aesthetic and functional visuals. 2) Improve speed/reproducibility of common tasks/pieces of code in scRNA-seq analysis with a single or group of functions. For citation please use: Marsh SE (2021) "Custom Visualizations &amp; Functions for Streamlined Analyses of Single Cell Sequencing" <a href="https://doi.org/10.5281/zenodo.5706430" class="external-link">doi:10.5281/zenodo.5706430</a>
 RRID:SCR_024675.</p>
    </div>


    <div id="package-options">
    <h2>Package options</h2>
    


<p>scCustomize uses the following <code><a href="https://rdrr.io/r/base/options.html" class="external-link">options()</a></code> to configure behavior:</p>
<dl><dt><code>scCustomize_warn_raster_iterative</code></dt>
<dd><p>Show message about setting <code>raster</code> parameter
in <code><a href="Iterate_FeaturePlot_scCustom.html">Iterate_FeaturePlot_scCustom</a></code> if <code>raster = FALSE</code> and <code>single_pdf = TRUE</code>
due to large file sizes.</p></dd>

<dt><code>scCustomize_warn_raster_LIGER</code></dt>
<dd><p>Show warning about rasterization of points in
<code><a href="DimPlot_LIGER.html">DimPlot_LIGER</a></code> due to new functionality compared to LIGER.</p></dd>

<dt><code>scCustomize_warn_na_cutoff</code></dt>
<dd><p>Show message about properly setting <code>na_cutoff</code> parameter in <code><a href="FeaturePlot_scCustom.html">FeaturePlot_scCustom</a></code>.</p></dd>

#'   <dt><code>scCustomize_warn_zero_na_cutoff</code></dt>
<dd><p>Show message about properly setting <code>na_cutoff</code> parameter in <code><a href="FeaturePlot_scCustom.html">FeaturePlot_scCustom</a></code> if <code>na_cutoff</code> is set to exactly zero.</p></dd>

<dt><code>scCustomize_warn_vln_raster_iterative</code></dt>
<dd><p>Show message about <code><a href="Iterate_VlnPlot_scCustom.html">Iterate_VlnPlot_scCustom</a></code>
when <code>pt.size &gt; 0</code> due to current lack of raster support in <code><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></code></p></dd>

<dt><code>scCustomize_warn_LIGER_dim_labels</code></dt>
<dd><p>Show message about <code><a href="DimPlot_LIGER.html">DimPlot_LIGER</a></code>
parameter <code>reduction_label</code> as LIGER objects do not store dimensionality reduction name and
and therefore needs to be set manually.</p></dd>

<dt><code>scCustomize_warn_DimPlot_split_type</code></dt>
<dd><p>Show message about <code><a href="DimPlot_scCustom.html">DimPlot_scCustom</a></code>
parameter <code>split.by</code> and <code>split_seurat</code> to alert user to difference in returned plots between
scCustomize and Seurat.</p></dd>

<dt><code>scCustomize_warn_FeatureScatter_split_type</code></dt>
<dd><p>Show message about <code><a href="FeatureScatter_scCustom.html">FeatureScatter_scCustom</a></code>
parameter <code>split.by</code> and <code>split_seurat</code> to alert user to difference in returned plots between
scCustomize and Seurat.</p></dd>

<dt><code>scCustomize_warn_LIGER_dim_labels_plotFactors</code></dt>
<dd><p>Show message about <code><a href="plotFactors_scCustom.html">plotFactors_scCustom</a></code>
parameter <code>reduction_label</code> as LIGER objects do not store dimensionality reduction name and
and therefore needs to be set manually.</p></dd>


</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://github.com/samuel-marsh/scCustomize" class="external-link">https://github.com/samuel-marsh/scCustomize</a></p></li>
<li><p><a href="https://samuel-marsh.github.io/scCustomize/">https://samuel-marsh.github.io/scCustomize/</a></p></li>
<li><p><a href="https://doi.org/10.5281/zenodo.5706431" class="external-link">doi:10.5281/zenodo.5706431</a></p></li>
<li><p>Report bugs at <a href="https://github.com/samuel-marsh/scCustomize/issues" class="external-link">https://github.com/samuel-marsh/scCustomize/issues</a></p></li>
</ul></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p><strong>Maintainer</strong>: Samuel Marsh <a href="mailto:sccustomize@gmail.com">sccustomize@gmail.com</a> (<a href="https://orcid.org/0000-0002-3012-6945" class="external-link">ORCID</a>)</p>
<p>Other contributors:</p><ul><li><p>Ming Tang <a href="mailto:tangming2005@gmail.com">tangming2005@gmail.com</a> [contributor]</p></li>
<li><p>Velina Kozareva [contributor]</p></li>
<li><p>Lucas Graybuck <a href="mailto:lucasg@alleninstitute.org">lucasg@alleninstitute.org</a> [contributor]</p></li>
<li><p>Zoe Clarke <a href="mailto:zoe.clarke@utoronto.ca">zoe.clarke@utoronto.ca</a> (<a href="https://orcid.org/0000-0003-0336-8049" class="external-link">ORCID</a>) [contributor]</p></li>
</ul></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Samuel Marsh.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

